<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="BiocAsia2023">
<title>Unlocking single cell spatial omics analyses with scdney - IMC • BiocAsia2023</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Unlocking single cell spatial omics analyses with scdney - IMC">
<meta property="og:description" content="BiocAsia2023">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">BiocAsia2023</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/VisiumVersion3.html">Unlocking single cell spatial omics analyses with scdney - Visium</a>
    <a class="dropdown-item" href="../articles/breastCancerIMC.html">Unlocking single cell spatial omics analyses with scdney - IMC</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="breastCancerIMC_files/htmlwidgets-1.6.2/htmlwidgets.js"></script><link href="breastCancerIMC_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="breastCancerIMC_files/datatables-binding-0.30/datatables.js"></script><link href="breastCancerIMC_files/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="breastCancerIMC_files/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="breastCancerIMC_files/dt-core-1.13.4/js/jquery.dataTables.min.js"></script><link href="breastCancerIMC_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="breastCancerIMC_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Unlocking single cell spatial omics analyses with scdney - IMC</h1>
                        <h4 data-toc-skip class="author">Yue Cao, Andy
Tran, Dario Strbenac, Nicholas Robertson Jean Yang</h4>
            
            <h4 data-toc-skip class="date">15 October, 2023</h4>
      
      
      <div class="d-none name"><code>breastCancerIMC.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## library(devtools)</span></span>
<span><span class="co">## library(BiocManager)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/scFeatures/" class="external-link">scFeatures</a></span><span class="op">)</span> <span class="co">## devtools::install_github("SydneyBioX/scFeatures")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/ClassifyR/" class="external-link">ClassifyR</a></span><span class="op">)</span> <span class="co">## BiocManager::install("ClassifyR", dependencies = TRUE)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/lisaClust/" class="external-link">lisaClust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jrnold/ggthemes" class="external-link">ggthemes</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">spicyR</span><span class="op">)</span> <span class="co">## BiocManager::install("spicyR")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com" class="external-link">plotly</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scattermore</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival" class="external-link">survival</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/survminer/index.html" class="external-link">survminer</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://spatstat.org/" class="external-link">spatstat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://had.co.nz/reshape" class="external-link">reshape</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># code for plotting purpose</span></span>
<span></span>
<span></span>
<span><span class="va">plot_boxplot</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span> <span class="va">feature</span> <span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">data_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span></span>
<span>  <span class="va">data_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape/man/melt-24.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">data_plot</span> <span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data_plot</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X1"</span>, <span class="st">"X2"</span>, <span class="st">"value"</span><span class="op">)</span></span>
<span>  <span class="va">data_plot</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span>  <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span> <span class="va">data_plot</span><span class="op">$</span><span class="va">X2</span><span class="op">)</span>, <span class="st">"_cond_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">data_plot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span> x <span class="op">=</span> <span class="va">X1</span>, y <span class="op">=</span> <span class="va">value</span> , colour <span class="op">=</span> <span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>   </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span>  </span>
<span></span>
<span><span class="va">plot_barplot</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">data</span> , <span class="va">dodge</span><span class="op">=</span><span class="cn">F</span>  <span class="op">)</span><span class="op">{</span></span>
<span></span>
<span>  <span class="va">data</span><span class="op">$</span><span class="va">patient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span> <span class="op">)</span>, <span class="st">"_cond_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">data</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span> <span class="op">)</span>, <span class="st">"_cond_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape/man/melt-24.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">data</span>, id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"patient"</span>, <span class="st">"condition"</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></span>
<span> </span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">data</span> , <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span> x <span class="op">=</span> <span class="va">patient</span> , y <span class="op">=</span> <span class="va">value</span> , fill <span class="op">=</span> <span class="va">variable</span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span>   </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span>   <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">condition</span>, scale<span class="op">=</span><span class="st">"free"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> </span>
<span> </span>
<span> <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span> <span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span> </span>
<span><span class="va">draw_dotplot</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">data_sce</span>, <span class="va">sample</span>, <span class="va">celltype</span> , <span class="va">group</span> , <span class="va">group_of_interest</span> <span class="op">)</span><span class="op">{</span></span>
<span></span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span> <span class="va">data_sce</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># for each patient, calculate the proportion of each cell type in each region.  </span></span>
<span></span>
<span>  <span class="va">df_plot</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span> <span class="va">thispatient</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">this_df</span> <span class="op">&lt;-</span> <span class="va">df</span><span class="op">[</span><span class="va">df</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span><span class="op">==</span> <span class="va">thispatient</span>, <span class="op">]</span></span>
<span>    <span class="va">temp_df</span> <span class="op">&lt;-</span>   <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span> <span class="va">this_df</span><span class="op">$</span><span class="va">region</span>, <span class="va">this_df</span><span class="op">[[</span><span class="va">celltype</span><span class="op">]</span><span class="op">]</span> <span class="op">)</span></span>
<span>    <span class="va">temp_df</span> <span class="op">&lt;-</span>  <span class="va">temp_df</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">temp_df</span><span class="op">)</span></span>
<span>     <span class="va">temp_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>  <span class="va">temp_df</span><span class="op">)</span></span>
<span>    <span class="va">temp_df</span><span class="op">$</span><span class="va">patient</span> <span class="op">&lt;-</span>  <span class="va">thispatient</span></span>
<span>    <span class="va">temp_df</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">[</span><span class="va">clinical</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="va">thispatient</span>,  <span class="va">group</span> <span class="op">]</span>  </span>
<span>    <span class="va">df_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">df_plot</span>, <span class="va">temp_df</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co"># for each region, calculate the average proportion of each cell type across all individuals </span></span>
<span>  <span class="va">df_plot</span> <span class="op">&lt;-</span> <span class="va">df_plot</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span> <span class="va">Var1</span> , <span class="va">Var2</span>, <span class="va">group</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mean_proportion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Freq</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>  <span class="co"># we are only interested in the short term and long term survival individuals </span></span>
<span>  <span class="va">df_plot</span>  <span class="op">&lt;-</span> <span class="va">df_plot</span> <span class="op">[</span> <span class="va">df_plot</span><span class="op">$</span><span class="va">group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">group_of_interest</span>, <span class="op">]</span></span>
<span>   </span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_plot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">Var2</span>, x <span class="op">=</span> <span class="va">Var1</span> ,colour <span class="op">=</span><span class="va">mean_proportion</span>  , size <span class="op">=</span> <span class="va">mean_proportion</span> <span class="op">)</span><span class="op">)</span><span class="op">+</span>  </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="op">~</span><span class="va">group</span>, scales <span class="op">=</span> <span class="st">"free"</span>, space <span class="op">=</span> <span class="st">"free"</span> <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Region"</span> <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Cell type"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">draw_region_clustering_result</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span> <span class="va">data_sce</span>, <span class="va">sample</span> , <span class="va">selected_sample</span>   <span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># get meta data for a selected patient to visualise </span></span>
<span>  <span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span></span>
<span>  <span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">metadata</span><span class="op">[</span><span class="va">metadata</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="va">selected_sample</span> ,  <span class="op">]</span></span>
<span>  <span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span></span>
<span>  </span>
<span>  <span class="va">plotlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># define the list to store images for region highlighting </span></span>
<span>  <span class="va">plotlist_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># define the list to store images for celltype highlighting </span></span>
<span>   </span>
<span>  <span class="co"># optional: define a colour palette</span></span>
<span>  <span class="co"># you can also specify your own colour to use in the variable color_codes </span></span>
<span>  <span class="va">tableau_palette</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/scale_color_tableau.html" class="external-link">scale_colour_tableau</a></span><span class="op">(</span> palette <span class="op">=</span> <span class="st">"Tableau 20"</span><span class="op">)</span> </span>
<span>  <span class="va">color_codes</span> <span class="op">&lt;-</span> <span class="va">tableau_palette</span><span class="op">$</span><span class="fu">palette</span><span class="op">(</span><span class="fl">18</span><span class="op">)</span></span>
<span>  <span class="va">color_codes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">color_codes</span>, <span class="st">"darkgrey"</span><span class="op">)</span>  <span class="co"># for all other regions apart from region of interest, make the colour grey </span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">color_codes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">description</span><span class="op">)</span> ,  <span class="st">"other regions"</span><span class="op">)</span></span>
<span>  </span>
<span>  </span>
<span>  <span class="co"># look through each region to highlight the region of interest, as well as the cell type in the region of interest</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span> <span class="va">thisregion</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">region</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>      </span>
<span>        <span class="co"># select the region of interest</span></span>
<span>        <span class="va">selected_region_index</span> <span class="op">&lt;-</span>  <span class="va">metadata</span><span class="op">$</span><span class="va">region</span> <span class="op">==</span> <span class="va">thisregion</span></span>
<span>        </span>
<span>        <span class="co"># for all other regions, define them as "other regions" so that they will be greyed out </span></span>
<span>        <span class="va">metadata</span><span class="op">$</span><span class="va">selected_region</span> <span class="op">&lt;-</span>  <span class="st">"other regions"</span></span>
<span>        <span class="va">metadata</span><span class="op">$</span><span class="va">selected_region</span><span class="op">[</span><span class="va">selected_region_index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"selected region"</span></span>
<span>        </span>
<span>        <span class="co"># for all cell types outside the region of the interest, also make them greyed out </span></span>
<span>        <span class="va">metadata</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">metadata</span><span class="op">$</span><span class="va">description</span></span>
<span>        <span class="va">metadata</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span><span class="op">!</span><span class="va">selected_region_index</span> <span class="op">]</span> <span class="op">&lt;-</span>   <span class="st">"other regions"</span></span>
<span>        <span class="va">metadata</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">celltype</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">description</span><span class="op">)</span>, <span class="st">"other regions"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>        <span class="co"># plot ggplot highlighting the region of interest</span></span>
<span>       <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span> , colour <span class="op">=</span> <span class="va">selected_region</span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">thisregion</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span> , <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span>         </span>
<span>       <span class="co"># plot ggplot highlighting the celltypes in the region of interest </span></span>
<span>       <span class="va">p2</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span> , colour <span class="op">=</span>  <span class="va">celltype</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.7</span> <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">thisregion</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span>  <span class="va">color_codes</span><span class="op">)</span></span>
<span>       </span>
<span>      <span class="va">plotlist</span> <span class="op">[[</span><span class="va">thisregion</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">p</span></span>
<span>       </span>
<span>      <span class="va">plotlist_celltype</span> <span class="op">[[</span><span class="va">thisregion</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">p2</span></span>
<span>      </span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plotlist</span> , ncol <span class="op">=</span> <span class="fl">5</span>, nrow <span class="op">=</span> <span class="fl">1</span> , common.legend <span class="op">=</span> <span class="cn">T</span> <span class="op">)</span></span>
<span>  <span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plotlist_celltype</span> , ncol <span class="op">=</span> <span class="fl">5</span>, nrow <span class="op">=</span> <span class="fl">1</span> , common.legend <span class="op">=</span> <span class="cn">T</span> <span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="va">draw_selected_region_boxplot</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">data_sce</span>, <span class="va">sample</span> , <span class="va">celltype</span>, <span class="va">group</span> ,  <span class="va">group_of_interest</span>, <span class="va">select_region</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>    <span class="va">df</span> <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">[</span>, <span class="va">sample</span> <span class="op">]</span>, <span class="op">]</span></span>
<span>   <span class="va">df</span><span class="op">$</span><span class="va">region</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">region</span></span>
<span>   <span class="va">df</span><span class="op">[[</span><span class="va">celltype</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[[</span><span class="va">celltype</span><span class="op">]</span><span class="op">]</span></span>
<span>  </span>
<span>   <span class="va">df_plot</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>   <span class="kw">for</span> <span class="op">(</span> <span class="va">thispatient</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="va">this_df</span> <span class="op">&lt;-</span> <span class="va">df</span><span class="op">[</span><span class="va">df</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="va">thispatient</span>, <span class="op">]</span></span>
<span>      <span class="va">temp_df</span> <span class="op">&lt;-</span>   <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span> <span class="va">this_df</span><span class="op">$</span><span class="va">region</span>, <span class="va">this_df</span><span class="op">[[</span><span class="va">celltype</span><span class="op">]</span><span class="op">]</span> <span class="op">)</span></span>
<span>      <span class="va">temp_df</span> <span class="op">&lt;-</span>  <span class="va">temp_df</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">temp_df</span><span class="op">)</span></span>
<span>      <span class="va">temp_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>  <span class="va">temp_df</span><span class="op">)</span></span>
<span>      <span class="va">temp_df</span><span class="op">$</span><span class="va">patient</span> <span class="op">&lt;-</span>  <span class="va">thispatient</span></span>
<span>      <span class="va">temp_df</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span> <span class="va">this_df</span><span class="op">[[</span><span class="va">group</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>      <span class="va">df_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">df_plot</span>, <span class="va">temp_df</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>     </span>
<span>  <span class="va">df_plot_region</span> <span class="op">&lt;-</span> <span class="va">df_plot</span><span class="op">[</span><span class="va">df_plot</span><span class="op">$</span><span class="va">Var1</span> <span class="op">==</span> <span class="va">select_region</span> , <span class="op">]</span></span>
<span>  </span>
<span>    </span>
<span>  <span class="va">df_plot_region</span>  <span class="op">&lt;-</span>  <span class="va">df_plot_region</span> <span class="op">[</span> <span class="va">df_plot_region</span><span class="op">$</span><span class="va">group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">group_of_interest</span>, <span class="op">]</span></span>
<span>   </span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_plot_region</span>,  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span>  <span class="va">Var2</span>,  y <span class="op">=</span> <span class="va">Freq</span>, colour <span class="op">=</span> <span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>  <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Proportion"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Cell type"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Selected region"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.25</span><span class="op">)</span></span>
<span>   </span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">p</span> <span class="op">)</span></span>
<span> </span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>As single cell technology advances, the recent development of spatial
omics allows us to examine the spatial organisation of cells within
tissues in their native environment. This workshop will discuss the
challenges and analytical focus associated with using multi-sample
spatial datasets for disease risk prognosis. We will also talk about
general analytical strategies and the critical thinking questions that
arise in the workflow.</p>
<br><div class="aimbox">
<div class="section level4">
<h4 class="unnumbered" id="preparation-and-assumed-knowledge">
<span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-pencil-alt fa-stack-1x fa-inverse"></i></span>
Preparation and assumed knowledge<a class="anchor" aria-label="anchor" href="#preparation-and-assumed-knowledge"></a>
</h4>
<ul>
<li>Knowledge of R syntax</li>
<li>Familiarity with the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment
class</a>
</li>
<li>Basic knowledge in <a href="https://bioconductor.org/books/release/OSCA/index.html" class="external-link">single
cell data analysis</a>. You can access our <a href="https://github.com/SydneyBioX/scdney#scdney-workshops-series" class="external-link">previous
workshops</a> for a quick review in single cell data analysis.</li>
<li>Ability to install all required R packages, please check
<code>sessionInfo</code> at the end of this document to ensure you are
using the correct version.</li>
<li>Familiarity with our previous workshop vignette on <a href="https://sydneybiox.github.io/BIS2019_SC/" class="external-link">Introduction to Single
Cell RNA-seq Analysis</a>
</li>
</ul>
</div>
<div class="section level4">
<h4 class="unnumbered" id="learning-objectives">
<span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-location-arrow fa-stack-1x fa-inverse"></i></span>
Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h4>
<ul>
<li>Describe and visualise spatial omics datasets<br>
</li>
<li>Calculate different measures that describe the spatial distribution
of cell types</li>
<li>Generate individual feature representations from a cell-level
expression matrix</li>
<li>Perform multi-view disease outcome prognosis with the package
<code>ClassifyR</code>
</li>
<li>Develop understanding on
<ul>
<li>how to assess the performance of classification and survival
models<br>
</li>
<li>how to identify and assess individual performance given a survival
model.</li>
</ul>
</li>
<li>Explore various strategies for disease outcome prognosis using
spatial omics data<br>
</li>
</ul>
</div>
</div>
<p><strong>Note:</strong> This data analysis workshop offers
participants the opportunity to engage in hands-on analysis using R.
However, if you are not comfortable with coding in R, you can still
acquire valuable interpretation skills by reviewing the output we
provide in this file. <br></p>
<div class="section level4">
<h4 class="unnumbered" id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h4>
<p>Structure of the 3-hour workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Exploring spatial data</td>
<td>35m</td>
</tr>
<tr class="odd">
<td>Feature engineering with scFeatures</td>
<td>50m</td>
</tr>
<tr class="even">
<td>Break</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Survival analysis with ClassifyR</td>
<td>45m</td>
</tr>
<tr class="even">
<td>Identify cohort heterogeneity</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Concluding remarks</td>
<td></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="initial-exploration-and-visualisation">Initial exploration and visualisation<a class="anchor" aria-label="anchor" href="#initial-exploration-and-visualisation"></a>
</h2>
<div class="section level4">
<h4 class="l-" id="data-and-background">Data and background<a class="anchor" aria-label="anchor" href="#data-and-background"></a>
</h4>
<p>The widely-known METABRIC breast cancer cohort has recently had
imaging mass cytometry (cell-level resolution) generated for a subset of
it. The publication describing this data is <a href="https://www.nature.com/articles/s43018-020-0026-6" class="external-link">Imaging Mass
Cytometry and Multiplatform Genomics Define the Phenogenomic Landscape
of Breast Cancer</a>, <em>Nature Cancer</em>, 2020. There are 483 cancer
samples with IMC data. However, the subset of interest is those
individuals who do not have lymph node metastasis. Can their risk of
recurrence accurately be estimated and therefore inform how aggressively
they need to be treated? The other component of the analysis is patient
clinical data, which has been sourced from Supplementary Table 5 of <a href="https://www.nature.com/articles/s41586-019-1007-8" class="external-link">Dynamics of
Breast-cancer Relapse Reveal Late-recurring ER-positive Genomic
Subgroups</a>, <em>Nature</em>, 2019. The original data has been
restricted to images with at least 400 cells, no lymph node cancer and
Stage 1.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Quick look at data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##           MB-0002:345:93 MB-0002:345:107 MB-0002:345:113 MB-0002:345:114</span></span>
<span><span class="co">## HH3_total     4.82032390       3.8229411      2.55845170      4.82208870</span></span>
<span><span class="co">## CK19          1.21185160       1.3223530      0.13832258      0.33366668</span></span>
<span><span class="co">## CK8_18        2.80274720       2.4720588      0.60545164      2.43351100</span></span>
<span><span class="co">## Twist         0.14651649       0.1176471      0.09877419      0.20791112</span></span>
<span><span class="co">## CD68          0.07863736       0.1016471      0.03225806      0.05237778</span></span>
<span><span class="co">##           MB-0002:345:125</span></span>
<span><span class="co">## HH3_total      2.41391110</span></span>
<span><span class="co">## CK19           0.07055555</span></span>
<span><span class="co">## CK8_18         0.20724444</span></span>
<span><span class="co">## Twist          0.12004445</span></span>
<span><span class="co">## CD68           0.00000000</span></span></code></pre>
<p>Basic characteristics of the data objects:<br>
- The dataset contains 38 proteins and 76307 cells.<br>
- The outcome is recurrence-free survival.</p>
</div>
<div class="section level3">
<h3 id="exploration-1-how-complex-is-my-data">Exploration 1: How complex is my data?<a class="anchor" aria-label="anchor" href="#exploration-1-how-complex-is-my-data"></a>
</h3>
<p>At the start of any analysis pipeline, it is often good to explore
the data to get a sense of the complexity. We usually do this by
exploring the distribution of the outcomes and various variables in the
individuals’ meta-data. Here, we use cross-tabulation to examine the
following variables:</p>
<ul>
<li>Surgery vs death</li>
<li>ER status</li>
<li>Grade</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Stage and death"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Stage and death"</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">$</span><span class="va">Breast.Surgery</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">Death</span>, useNA <span class="op">=</span> <span class="st">"ifany"</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">##                    </span></span>
<span><span class="co">##                      0  1</span></span>
<span><span class="co">##   BREAST CONSERVING 38 14</span></span>
<span><span class="co">##   MASTECTOMY        16  6</span></span>
<span><span class="co">##   &lt;NA&gt;               2  1</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Number of individuals based on ER status"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Number of individuals based on ER status"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">$</span><span class="va">ER.Status</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## neg pos </span></span>
<span><span class="co">##  12  64</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Number of individuals based on Grade"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Number of individuals based on Grade"</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">$</span><span class="va">Grade</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  1  2  3 </span></span>
<span><span class="co">## 14 30 28</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="exploration-2-how-to-visualise-my-data">Exploration 2: How to visualise my data?<a class="anchor" aria-label="anchor" href="#exploration-2-how-to-visualise-my-data"></a>
</h3>
<p>Typically in single-cell data analysis, we perform dimension
reduction to project the high dimensional cell x gene matrix on to 2D
space. This allows us to visualise various things of interest, such as
distribution of cell types and disease outcomes. In this dataset, cells
were classified into 22 cell types based on their markers.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">data_sce</span>, scale<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># With the UMAP function we can highlight by meta data of interest</span></span>
<span><span class="co"># Here we highlight the cell types and sample ID</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">data_sce</span>, colour_by <span class="op">=</span> <span class="st">"description"</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">data_sce</span>, colour_by <span class="op">=</span> <span class="st">"metabricId"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span> plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-6-1.png" width="1152"></p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i>
<i class="far fa-chart-bar fa-stack-1x fa-inverse"></i> </span>
<strong>Interactive Q&amp;A:</strong></p>
<p>What can we learn from these illustrations? Is there anything
interesting in the plot? Questions to consider include:</p>
<ul>
<li>Q1: Does each cell type cluster together?</li>
<li>Q2: When there is a large number of categories, are dimensionality
reduction plots interpretable or misleading due to overplotting?</li>
</ul>
</div>
<p>Visualise selected patient:<br>
Here we selectively visualise two individuals by highlighting only the
selected individuals.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract the meta data and the UMAP dimension of the dataset</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">data_sce</span>, <span class="st">"UMAP"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># For the selected patient MB-0475, denote it as "selected patient" </span></span>
<span><span class="co"># and all other individuals as "other individuals"</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">selected_patient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0475"</span>, <span class="st">"seleted patient"</span> , <span class="st">"other individuals"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate a UMAP plot and store it in R object `a`   </span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span><span class="va">UMAP1</span> , y <span class="op">=</span> <span class="va">UMAP2</span> , colour <span class="op">=</span> <span class="va">selected_patient</span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/scattermore/man/geom_scattermore.html" class="external-link">geom_scattermore</a></span><span class="op">(</span>pointsize <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span> , <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span> </span>
<span><span class="co"># For the selected patient MB-0628, denote it as "selected patient" </span></span>
<span><span class="co"># and all other individuals as "other individuals"</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">selected_patient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span> <span class="va">metadata</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0628"</span>, <span class="st">"seleted patient"</span> , <span class="st">"other individuals"</span><span class="op">)</span></span>
<span>        </span>
<span><span class="co"># Generate a UMAP plot and store it in R object `b`   </span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span><span class="va">UMAP1</span> , y <span class="op">=</span> <span class="va">UMAP2</span> , colour <span class="op">=</span> <span class="va">selected_patient</span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/scattermore/man/geom_scattermore.html" class="external-link">geom_scattermore</a></span><span class="op">(</span>pointsize <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span> , <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span> </span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span> <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-7-1.png" width="1152"></p>
</div>
<div class="section level3">
<h3 class="tabset" id="exploration-3-is-there-a-spatial-structure-in-my-data">Exploration 3: Is there a spatial structure in my
data?<a class="anchor" aria-label="anchor" href="#exploration-3-is-there-a-spatial-structure-in-my-data"></a>
</h3>
<p>The advantage with spatial omics is that we can examine the
organisation of the cell types as it occurs on the tissue slide. Here,
we visualise one of the slides from a patient. As an optional exercise,
you can: - permute the cell type label.<br>
- permute the spatial coordinate.<br>
to give a sense of what is random ordering.</p>
<div class="section level4">
<h4 class="unnumbered" id="spatial-plot">Spatial plot<a class="anchor" aria-label="anchor" href="#spatial-plot"></a>
</h4>
<p>We select a particular patient “MB-0263” and visualise its spatial
pattern using ggplot.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># obtaining the meta data for this patient </span></span>
<span><span class="va">one_sample</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span>, <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span>  <span class="op">==</span> <span class="st">"MB-0263"</span><span class="op">]</span></span>
<span><span class="va">one_sample</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">one_sample</span><span class="op">)</span></span>
<span><span class="va">one_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">one_sample</span><span class="op">)</span></span>
<span> </span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">one_sample</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span>, y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span> <span class="va">description</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/scale_color_tableau.html" class="external-link">scale_colour_tableau</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Original slide"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">a</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-8-1.png" width="576"></p>
</div>
<div class="section level4">
<h4 class="unnumbered" id="optional-code-random-spatial-pattern">[Optional code] Random spatial pattern<a class="anchor" aria-label="anchor" href="#optional-code-random-spatial-pattern"></a>
</h4>
<p>The code here investigates permutations of spatially resolved data.
Please examine the next tab for actual results.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># "Optional: Permute the cell type labels"</span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">description_permute</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">one_sample</span><span class="op">$</span><span class="va">description</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">one_sample</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span><span class="va">description_permute</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/scale_color_tableau.html" class="external-link">scale_colour_tableau</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Permute the cell type label"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># "Optional: Permute the spatial coordinate"</span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">Location_Center_X_permute</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">one_sample</span><span class="op">$</span><span class="va">Location_Center_X</span><span class="op">)</span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">Location_Center_Y_permute</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">one_sample</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span></span>
<span><span class="va">c</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">one_sample</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X_permute</span> , y <span class="op">=</span> <span class="va">Location_Center_Y_permute</span>, colour <span class="op">=</span> <span class="va">description</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/scale_color_tableau.html" class="external-link">scale_colour_tableau</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Permute the X, Y coordinate"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 class="unnumbered" id="spatial-structure">Spatial structure<a class="anchor" aria-label="anchor" href="#spatial-structure"></a>
</h4>
<p>The aim here is to have an understanding of the concept of spatial
randomness. Spatial statistics is a topic of study that encompasses a
wide range of research. The next two code chucks focus on the
examination of two distinct permutation strategies. The objective of
this investigation is to gain an understanding of how various
permutation strategies might yield varied perceptions of randomness.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## ggarrange provides a way to arrange multiple ggplots for efficient visual comparison</span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span> plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>, <span class="va">b</span>, <span class="va">c</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> </span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-11-1.png" width="1152"></p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i>
<i class="far fa-chart-bar fa-stack-1x fa-inverse"></i> </span>
<strong>Interactive Q&amp;A:</strong></p>
<p>Q3: Is there a structure in the data or is the cell type randomly
distribution?</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="describing-tissue-microenvrionments-and-cellular-neighbourhoods">Describing tissue microenvrionments and cellular neighbourhoods<a class="anchor" aria-label="anchor" href="#describing-tissue-microenvrionments-and-cellular-neighbourhoods"></a>
</h2>
<p>Spatial data allow for the identification of a variety of
characteristics including distinct cell types within an image, providing
an overview of the tissue environment. This allows scientists to explore
the cellular architecture and environment and its association with
phenotype information (e.g meta-data). For our data story, we are
interested in whether the individuals have a good or poor outcome. The
outcome is often called a ‘prognosis’ and a good outcome is sometimes
called ‘favourable’. In this section, we examine graphically how
cell-type co-localisation varies across spatial regions and how is such
information associated with individual survival outcome.</p>
<div class="section level3">
<h3 id="do-cell-type-co-localise-in-specfic-regions">Do cell type co-localise in specfic regions?<a class="anchor" aria-label="anchor" href="#do-cell-type-co-localise-in-specfic-regions"></a>
</h3>
<p>We begin by examining how can we identify and visualise regions of
tissue where spatial associations between cell-types are similar. There
are many packages that perform this task and here we use the <a href="https://www.bioconductor.org/packages/devel/bioc/html/lisaClust.html" class="external-link">lisaClust
function</a> that is based on “local L-function” to spatially cluster
cells into different regions with similar cell type composition.</p>
<p style="color:blue">
<em>This has already been pre-loaded for you. Please proceed to the next
task.</em>
</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="va">BPPARAM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">16</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Cluster cells into spatial regions with similar composition.</span></span>
<span><span class="va">data_sce</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/lisaClust.html" class="external-link">lisaClust</a></span><span class="op">(</span></span>
<span>  <span class="va">data_sce</span> ,</span>
<span>  k <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  Rs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>  sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Location_Center_X"</span>, <span class="st">"Location_Center_Y"</span><span class="op">)</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"description"</span>,</span>
<span>  imageID <span class="op">=</span> <span class="st">"ImageNumber"</span> ,</span>
<span>  regionName <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 class="tabset" id="which-regions-appear-to-be-different-between-good-and-poor-prognosis">Which regions appear to be different between good and
poor prognosis?<a class="anchor" aria-label="anchor" href="#which-regions-appear-to-be-different-between-good-and-poor-prognosis"></a>
</h3>
<p>We have use the <code>lisaClust</code> function in the previous
subsection to cluster cells into five different spatial regions. Next,
as a case study, we will compare individuals with good or poor prognosis
and examine graphically, if any regions appear to be different between
good or poor prognosis. We define:<br>
- Good prognosis as individuals with &gt; 10 years recurrence-free
survival and - Poor prognosis as individuals with &lt; 5 years
recurrence-free survival.</p>
<p>using the code below and store the information under
<code>clinical$survivalgroup</code>. The recurrence free survival (RFS)
is provided in the variable <code>timeRFS</code></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract time to recurrence-free survival</span></span>
<span><span class="va">clinical</span><span class="op">$</span><span class="va">survivalgroup</span> <span class="op">&lt;-</span> <span class="st">"neither"</span></span>
<span></span>
<span><span class="co">## Define poor and good prognosis</span></span>
<span><span class="va">clinical</span><span class="op">$</span><span class="va">survivalgroup</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span> <span class="va">clinical</span><span class="op">$</span><span class="va">timeRFS</span>  <span class="op">&lt;</span> <span class="fl">5</span><span class="op">*</span> <span class="fl">365</span><span class="op">)</span> <span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"poor"</span> </span>
<span><span class="va">clinical</span><span class="op">$</span><span class="va">survivalgroup</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span> <span class="va">clinical</span><span class="op">$</span><span class="va">timeRFS</span>  <span class="op">&gt;</span> <span class="fl">10</span><span class="op">*</span> <span class="fl">365</span><span class="op">)</span> <span class="op">]</span> <span class="op">&lt;-</span>  <span class="st">"good"</span></span></code></pre></div>
<div class="section level4">
<h4 class="unnumbered" id="individual-level">Individual level<a class="anchor" aria-label="anchor" href="#individual-level"></a>
</h4>
<p>Here we visualise the spatial domain (regions) detection result based
on one individual. We can either visualise all regions in one graph or
highlighting each region in separate graphs. Here we will use the
terminology “spatial domain” and “regions” interchangeably.</p>
<p>Depending on the number of regions, it may be more useful to
visualise the spatial regions either collectively in a single graph or
separately in multiple graphs. To visualise it in a single graph, the
<code><a href="https://rdrr.io/pkg/lisaClust/man/hatchingPlot.html" class="external-link">hatchingPlot()</a></code> function is used to create hatching patterns
for representating spatial regions and cell-types. The hatching geom is
used to create hatching patterns for representation of spatial
regions.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## To visualise it in a single graph</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/hatchingPlot.html" class="external-link">hatchingPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">data_sce</span>,</span>
<span>  region <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>  imageID <span class="op">=</span> <span class="st">"metabricId"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"description"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Location_Center_X"</span>, <span class="st">"Location_Center_Y"</span><span class="op">)</span> <span class="op">)</span> </span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-14-1.png" width="576"></p>
<p>We have written a small function
<code>draw_region_clustering_result</code> to visualise the data
separately in multiple graphs for the individual
<code>MB-0628</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw_region_clustering_result</span><span class="op">(</span><span class="va">data_sce</span> , </span>
<span>                              sample <span class="op">=</span> <span class="st">"metabricId"</span> , </span>
<span>                              selected_sample <span class="op">=</span>  <span class="st">"MB-0628"</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span></code></pre>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-15-1.png" width="960"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span></code></pre>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-15-2.png" width="960"></p>
</div>
<div class="section level4">
<h4 class="unnumbered" id="across-individuals">Across individuals<a class="anchor" aria-label="anchor" href="#across-individuals"></a>
</h4>
<p>We can better interpret the region output by summarising the
proportion of each cell type in a region across the individuals. We look
at the composition of cell types in each region, and compare between
prognostic outcomes.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw_dotplot</span><span class="op">(</span><span class="va">data_sce</span>,  </span>
<span>             sample <span class="op">=</span> <span class="st">"metabricId"</span> , </span>
<span>             celltype <span class="op">=</span> <span class="st">"description"</span> ,  </span>
<span>             group<span class="op">=</span> <span class="st">"survivalgroup"</span> , </span>
<span>             group_of_interest <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"poor"</span> , <span class="st">"good"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-16-1.png" width="960"></p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i>
<i class="far fa-chart-bar fa-stack-1x fa-inverse"></i> </span>
<strong>Interactive Q&amp;A:</strong></p>
<p>Q4: Which regions appear to be different between poor prognosis
(short-term survival) and good prognosis (long-term survival)
individuals?</p>
</div>
</div>
<div class="section level4">
<h4 class="unnumbered" id="further-exploration-by-visualising-selected-regions">Further exploration by visualising selected
regions<a class="anchor" aria-label="anchor" href="#further-exploration-by-visualising-selected-regions"></a>
</h4>
<p>The number of sub-cell types increase considerably when we want to
add spatial domain (region) information. To enhance clarity and
facilitate understanding, it may be helpful to choose a predetermined
region. The code generates a set of boxplots that enable the comparison
of cell type proportions between individuals with good and
poor prognosis in <code>region_5</code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw_selected_region_boxplot</span><span class="op">(</span><span class="va">data_sce</span>, </span>
<span>                             sample <span class="op">=</span> <span class="st">"metabricId"</span> , </span>
<span>                             celltype <span class="op">=</span><span class="st">"description"</span> , </span>
<span>                             group  <span class="op">=</span> <span class="st">"survivalgroup"</span>, </span>
<span>                             group_of_interest <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"poor"</span> , <span class="st">"good"</span><span class="op">)</span>, </span>
<span>                             select_region <span class="op">=</span> <span class="st">"region_5"</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-17-1.png" width="1152"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="how-do-we-generate-a-molecular-representation-for-each-individual">How do we generate a molecular representation for each
individual?<a class="anchor" aria-label="anchor" href="#how-do-we-generate-a-molecular-representation-for-each-individual"></a>
</h2>
<p>In this workshop, we use the scFeatures package to generate a
molecular representation for each individuals from the matrix of
<code>features x cells</code>. The molecular representation is
interpretable and hence facilitates downstream analysis of the
individual. In general, <code>scFeatures</code> generates features
across six categories representing different molecular views of cellular
characteristics. These include:</p>
<ol style="list-style-type: lower-roman">
<li>cell type proportions</li>
<li>cell type specific gene expressions</li>
<li>cell type specific pathway expressions</li>
<li>cell type specific cell-cell interaction (CCI) scores</li>
<li>overall aggregated gene expressions</li>
<li>spatial metrics</li>
</ol>
<p>The different types of features constructed enable a more
comprehensive multi-view understanding of each individual from a matrix
of <code>spot x cells</code>. By default, the function will generate a
total of 13 different types of features (feature_types) shown below and
store them in a list. All generated feature types are stored in a matrix
of <code>samples x features</code>.</p>
<pre><code><span><span class="co">##  [1] "proportion_raw"       "proportion_logit"     "proportion_ratio"    </span></span>
<span><span class="co">##  [4] "gene_mean_celltype"   "gene_prop_celltype"   "gene_cor_celltype"   </span></span>
<span><span class="co">##  [7] "gene_mean_bulk"       "gene_prop_bulk"       "gene_cor_bulk"       </span></span>
<span><span class="co">## [10] "L_stats"              "celltype_interaction" "morans_I"            </span></span>
<span><span class="co">## [13] "nn_correlation"</span></span></code></pre>
<p>In additional to the spatial measure, we can also consider region
information as an additional layer of information. Thus, the following
code aims to create cell-type specific features for each region. We use
the function “paste0” to construct <strong>region-specific sub-cell
types</strong> and name it as <code>celltype</code> in the R object
<code>data_sce</code>. For simplicity, in this workshop, the variable
<code>celltype</code> in the R object <code>data_sce</code> refers to
region-specific sub-cell types.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">region</span></span>
<span></span>
<span><span class="co"># Define a series of sub-cell types that is regional specific</span></span>
<span><span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">description</span> , <span class="st">"-"</span> , <span class="va">region</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Nnumber of cells in each sample"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Nnumber of cells in each sample"</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">)</span><span class="op">)</span> , options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-34850a6ab38cbf60b51c" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-34850a6ab38cbf60b51c">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77"],["MB-0002","MB-0064","MB-0128","MB-0130","MB-0132","MB-0136","MB-0138","MB-0142","MB-0145","MB-0154","MB-0168","MB-0175","MB-0180","MB-0192","MB-0201","MB-0208","MB-0225","MB-0227","MB-0232","MB-0240","MB-0245","MB-0246","MB-0249","MB-0252","MB-0255","MB-0256","MB-0258","MB-0260","MB-0263","MB-0275","MB-0308","MB-0318","MB-0320","MB-0347","MB-0351","MB-0367","MB-0377","MB-0383","MB-0394","MB-0397","MB-0400","MB-0401","MB-0405","MB-0410","MB-0420","MB-0439","MB-0451","MB-0454","MB-0455","MB-0461","MB-0463","MB-0467","MB-0470","MB-0475","MB-0480","MB-0481","MB-0487","MB-0498","MB-0508","MB-0516","MB-0518","MB-0520","MB-0531","MB-0536","MB-0571","MB-0584","MB-0591","MB-0596","MB-0604","MB-0605","MB-0628","MB-0635","MB-0636","MB-0642","MB-0661","MB-0662","MB-0663"],[1111,1284,1005,870,1115,1648,509,744,982,338,819,1002,1247,808,580,625,1036,1357,847,1026,1282,859,474,794,932,801,499,1016,755,418,1116,1380,847,789,803,552,1026,773,1201,1275,1138,923,819,1138,532,1307,913,1195,605,1630,1592,886,865,1742,593,1685,967,1338,1226,1086,662,1393,607,1023,681,1078,940,1679,1182,634,1730,836,524,1255,1553,1238,567]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Var1<\/th>\n      <th>Freq<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"className":"dt-right","targets":2},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script><div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Number of cells in each region-specific sub-cell type"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Number of cells in each region-specific sub-cell type"</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span><span class="op">)</span> , options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-c1ba4485dbf0155cf3b7" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-c1ba4485dbf0155cf3b7">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110"],["B cells-region1","B cells-region2","B cells-region3","B cells-region4","B cells-region5","Basal CKlow-region1","Basal CKlow-region2","Basal CKlow-region3","Basal CKlow-region4","Basal CKlow-region5","Endothelial-region1","Endothelial-region2","Endothelial-region3","Endothelial-region4","Endothelial-region5","Fibroblasts CD68+-region1","Fibroblasts CD68+-region2","Fibroblasts CD68+-region3","Fibroblasts CD68+-region4","Fibroblasts CD68+-region5","Fibroblasts-region1","Fibroblasts-region2","Fibroblasts-region3","Fibroblasts-region4","Fibroblasts-region5","HER2+-region1","HER2+-region2","HER2+-region3","HER2+-region4","HER2+-region5","HR- CK7--region1","HR- CK7--region2","HR- CK7--region3","HR- CK7--region4","HR- CK7--region5","HR- CK7+-region1","HR- CK7+-region2","HR- CK7+-region3","HR- CK7+-region4","HR- CK7+-region5","HR- CKlow CK5+-region1","HR- CKlow CK5+-region2","HR- CKlow CK5+-region3","HR- CKlow CK5+-region4","HR- CKlow CK5+-region5","HR- Ki67+-region1","HR- Ki67+-region2","HR- Ki67+-region3","HR- Ki67+-region4","HR- Ki67+-region5","HR+ CK7- Ki67+-region1","HR+ CK7- Ki67+-region2","HR+ CK7- Ki67+-region3","HR+ CK7- Ki67+-region4","HR+ CK7- Ki67+-region5","HR+ CK7- Slug+-region1","HR+ CK7- Slug+-region2","HR+ CK7- Slug+-region3","HR+ CK7- Slug+-region4","HR+ CK7- Slug+-region5","HR+ CK7--region1","HR+ CK7--region2","HR+ CK7--region3","HR+ CK7--region4","HR+ CK7--region5","HRlow CKlow-region1","HRlow CKlow-region2","HRlow CKlow-region3","HRlow CKlow-region4","HRlow CKlow-region5","Hypoxia-region1","Hypoxia-region2","Hypoxia-region3","Hypoxia-region4","Hypoxia-region5","Macrophages Vim+ CD45low-region1","Macrophages Vim+ CD45low-region2","Macrophages Vim+ CD45low-region3","Macrophages Vim+ CD45low-region4","Macrophages Vim+ CD45low-region5","Macrophages Vim+ Slug--region1","Macrophages Vim+ Slug--region2","Macrophages Vim+ Slug--region3","Macrophages Vim+ Slug--region4","Macrophages Vim+ Slug--region5","Macrophages Vim+ Slug+-region1","Macrophages Vim+ Slug+-region2","Macrophages Vim+ Slug+-region3","Macrophages Vim+ Slug+-region4","Macrophages Vim+ Slug+-region5","Myoepithelial-region1","Myoepithelial-region2","Myoepithelial-region3","Myoepithelial-region4","Myoepithelial-region5","Myofibroblasts-region1","Myofibroblasts-region2","Myofibroblasts-region3","Myofibroblasts-region4","Myofibroblasts-region5","T cells-region1","T cells-region2","T cells-region3","T cells-region4","T cells-region5","Vascular SMA+-region1","Vascular SMA+-region2","Vascular SMA+-region3","Vascular SMA+-region4","Vascular SMA+-region5"],[5,111,11,157,13,125,132,104,574,28,20,253,24,238,54,146,829,42,843,80,842,3191,296,3894,400,149,97,26,230,42,1924,602,667,2557,824,506,443,2823,2635,863,50,128,6,155,7,117,148,61,955,30,293,158,280,1128,453,58,39,28,789,147,1652,1719,1775,9299,8606,1742,353,86,791,271,77,165,64,616,23,17,230,19,275,38,15,200,16,284,26,49,140,37,372,88,18,262,261,700,150,714,4035,510,4774,722,72,1161,117,1138,99,15,256,27,363,38]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Var1<\/th>\n      <th>Freq<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"className":"dt-right","targets":2},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script><div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span> <strong>
Discussion:</strong></p>
<p>Q6: Are there any samples or cell types you would like to remove from
the data?</p>
</div>
<div class="section level3">
<h3 class="tabset" id="how-to-create-molecular-representations-of-individuals">How to create molecular representations of
individuals?<a class="anchor" aria-label="anchor" href="#how-to-create-molecular-representations-of-individuals"></a>
</h3>
<p>There are different ways you can use <code>scFeatures</code> to
generate molecular representations for individuals and it requires the
following information for spatial data.</p>
<ul>
<li>data,<br>
</li>
<li>sample,</li>
<li>X coordinates,</li>
<li>Y coordinates,</li>
<li>feature_types, and</li>
<li>type</li>
</ul>
<p>There are a total of 13 different types of features (feature_types)
that you can choose to generate. The argument type refers the type of
input data we have. This is either <code>scrna</code> (single-cell
RNA-sequencing data), <code>spatial_p</code> (spatial proteomics data),
or <code>spatial_t</code> (single cell spatial data).</p>
<div class="section level4">
<h4 class="unnumbered" id="calculate-feature_types">Calculate feature_types<a class="anchor" aria-label="anchor" href="#calculate-feature_types"></a>
</h4>
<p>Suppose that we are interested in determining the proportion of each
cell type in each individual within each region. It is necessary to
specify <code>type = spatial_p</code> to reflect that we have spatial
proteomics data and <code>feature_types = proportion_raw</code> to
indicate we intend to calculate cell type proportion for each of the
region-specific sub-cell types.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## [A] The next few lines extract specific information from data_sce as input to scFeatures. </span></span>
<span><span class="co">## Extract the expression matrix from data_sce</span></span>
<span><span class="va">IMCmatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract the sample information </span></span>
<span><span class="co">## append the condition to the individuals so we can easily retrieve the individuals condition </span></span>
<span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> </span>
<span><span class="va">cond</span>  <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">metabricId</span><span class="op">)</span>, <span class="op">]</span><span class="op">$</span><span class="va">survivalgroup</span></span>
<span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"_cond_"</span>, <span class="va">cond</span> <span class="op">)</span> </span>
<span></span>
<span><span class="co">## Extract the region-specific sub-cell types</span></span>
<span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span></span>
<span></span>
<span><span class="co">## Extract the spatial coordinates</span></span>
<span><span class="va">spatialCoords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">Location_Center_X</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### [B] Running scFeatures</span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/scFeatures/reference/scFeatures.html" class="external-link">scFeatures</a></span><span class="op">(</span> <span class="va">IMCmatrix</span>, </span>
<span>                                 sample <span class="op">=</span> <span class="va">sample</span>, </span>
<span>                                 celltype <span class="op">=</span> <span class="va">celltype</span>, </span>
<span>                                 spatialCoords <span class="op">=</span> <span class="va">spatialCoords</span>,</span>
<span>                                 feature_types <span class="op">=</span> <span class="st">"proportion_raw"</span>, </span>
<span>                                 type <span class="op">=</span> <span class="st">"spatial_p"</span> <span class="op">)</span></span></code></pre></div>
<p>The generated feature is a matrix of <code>samples x features</code>,
the code below shows the first 5 features and the first 5 individuals
for the region-specific sub-cell types.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scfeatures_result</span><span class="op">$</span><span class="va">proportion_raw</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                      B cells-region1 B cells-region2 B cells-region3</span></span>
<span><span class="co">## MB-0002_cond_neither    0.0000000000     0.000000000               0</span></span>
<span><span class="co">## MB-0064_cond_neither    0.0000000000     0.000000000               0</span></span>
<span><span class="co">## MB-0128_cond_good       0.0009950249     0.044776119               0</span></span>
<span><span class="co">## MB-0130_cond_good       0.0000000000     0.000000000               0</span></span>
<span><span class="co">## MB-0132_cond_neither    0.0000000000     0.007174888               0</span></span>
<span><span class="co">##                      B cells-region4 B cells-region5</span></span>
<span><span class="co">## MB-0002_cond_neither      0.00000000     0.000000000</span></span>
<span><span class="co">## MB-0064_cond_neither      0.00000000     0.000000000</span></span>
<span><span class="co">## MB-0128_cond_good         0.06666667     0.007960199</span></span>
<span><span class="co">## MB-0130_cond_good         0.00000000     0.000000000</span></span>
<span><span class="co">## MB-0132_cond_neither      0.00000000     0.000000000</span></span></code></pre>
</div>
<div class="section level4">
<h4 class="unnumbered" id="focus-sub-cell-type">Focus sub-cell type<a class="anchor" aria-label="anchor" href="#focus-sub-cell-type"></a>
</h4>
<p>Suppose we are only interested in the molecular representation for
individuals r <code>HR- CK7+</code> only within region 5.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## [A] The next few lines extract specific information from data_sce as input to scFeatures. </span></span>
<span><span class="co">## Select the HR- CK7+-region sub-cell type </span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span>   <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"HR- CK7+-region"</span> , <span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span>, fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">selected_data</span>  <span class="op">&lt;-</span>  <span class="va">IMCmatrix</span><span class="op">[</span>, <span class="va">index</span><span class="op">]</span></span>
<span><span class="va">selected_sample</span> <span class="op">&lt;-</span>  <span class="va">sample</span><span class="op">[</span><span class="va">index</span><span class="op">]</span></span>
<span><span class="va">selected_celltype</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span> <span class="va">index</span><span class="op">]</span> </span>
<span><span class="va">selected_spatialCoords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">Location_Center_X</span><span class="op">[</span><span class="va">index</span><span class="op">]</span>, </span>
<span>                               <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">[</span><span class="va">index</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### [B] Running scFeatures</span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/scFeatures/reference/scFeatures.html" class="external-link">scFeatures</a></span><span class="op">(</span> <span class="va">selected_data</span>, </span>
<span>                                 sample <span class="op">=</span> <span class="va">selected_sample</span>, </span>
<span>                                 celltype <span class="op">=</span> <span class="va">selected_celltype</span>,</span>
<span>                                 spatialCoords <span class="op">=</span> <span class="va">selected_spatialCoords</span>,</span>
<span>                                 feature_types <span class="op">=</span> <span class="st">"proportion_raw"</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span> <span class="op">)</span></span>
<span></span>
<span><span class="co">### [C] Visualize the regional composition makeup for each individual for HR- CK7+ and HR- CK7-</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">proportion_raw</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span></span>
<span><span class="fu">plot_barplot</span><span class="op">(</span> <span class="va">feature</span> <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Proportion raw feature"</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-22-1.png" width="1920"></p>
</div>
<div class="section level4">
<h4 class="unnumbered" id="all-cell-types-and-features-in-one-line-of-code">All cell types and features in one line of
code<a class="anchor" aria-label="anchor" href="#all-cell-types-and-features-in-one-line-of-code"></a>
</h4>
<p>The code below enable us to generate all feature types for all cell
types in a line. Due to limitations with today’s computational capacity,
** Please DO NOT run it in today’s workshop, it will crash your
system**.</p>
<pre><code><span><span class="co"># here, we specify that this is a spatial proteomics data</span></span>
<span><span class="co"># scFeatures support parallel computation to speed up the process </span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/scFeatures/reference/scFeatures.html" class="external-link">scFeatures</a></span><span class="op">(</span><span class="va">IMCmatrix</span>, </span>
<span>                                type <span class="op">=</span> <span class="st">"spatial_p"</span>,</span>
<span>                                sample <span class="op">=</span> <span class="va">sample</span>, </span>
<span>                                celltype <span class="op">=</span> <span class="va">celltype</span>, </span>
<span>                                spatialCoords <span class="op">=</span> <span class="va">spatialCoords</span>,</span>
<span>                                ncores <span class="op">=</span> <span class="fl">32</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="what-can-we-learn-from-the-outputs-from-scfeatures">What can we learn from the outputs from scFeatures?<a class="anchor" aria-label="anchor" href="#what-can-we-learn-from-the-outputs-from-scfeatures"></a>
</h3>
<p>Assuming you have already generated a collection of molecular
representation for individuals, please load the prepared RDS file
<code>scfeatures_result.RDS</code>. Again, you can remind yourself that
all generated feature types are stored in a matrix of
<code>samples x features</code>.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Upload pre-generated RDS file</span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/biocasia_data/scfeatures_result.RDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we have generated a total of 13 feature types</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "proportion_raw"       "proportion_logit"     "proportion_ratio"    </span></span>
<span><span class="co">##  [4] "gene_mean_celltype"   "gene_prop_celltype"   "gene_cor_celltype"   </span></span>
<span><span class="co">##  [7] "gene_mean_bulk"       "gene_prop_bulk"       "gene_cor_bulk"       </span></span>
<span><span class="co">## [10] "L_stats"              "celltype_interaction" "morans_I"            </span></span>
<span><span class="co">## [13] "nn_correlation"</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># What is the number of features that we have generated </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">scfeatures_result</span>, <span class="va">dim</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $proportion_raw</span></span>
<span><span class="co">## [1]  77 110</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $proportion_logit</span></span>
<span><span class="co">## [1]  77 110</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $proportion_ratio</span></span>
<span><span class="co">## [1]   77 5995</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_mean_celltype</span></span>
<span><span class="co">## [1]   77 4180</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_prop_celltype</span></span>
<span><span class="co">## [1]   77 4180</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_cor_celltype</span></span>
<span><span class="co">## [1]    77 27958</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_mean_bulk</span></span>
<span><span class="co">## [1] 77 38</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_prop_bulk</span></span>
<span><span class="co">## [1] 77 38</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_cor_bulk</span></span>
<span><span class="co">## [1]  77 703</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $L_stats</span></span>
<span><span class="co">## [1]   77 3842</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $celltype_interaction</span></span>
<span><span class="co">## [1]   77 4393</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $morans_I</span></span>
<span><span class="co">## [1] 77 38</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $nn_correlation</span></span>
<span><span class="co">## [1] 77 38</span></span></code></pre>
<div class="section level4">
<h4 id="association-study---can-we-identify-differential-expression-for-a-feature-of-interest">Association study - can we identify “differential expression” for a
feature of interest<a class="anchor" aria-label="anchor" href="#association-study---can-we-identify-differential-expression-for-a-feature-of-interest"></a>
</h4>
<p>The R object <code>scfeatures_result</code> contains a variety of
features. A important question focuses on the identification of features
that reflect an association with the prognostic outcome, specifically
distinguishing between good and poor outcomes. The code provided below
demonstrates the use of the <code>limma()</code> function to fit a
linear model for the purpose of analysing gene_mean_celltype as an
illustration feature. The feature type known as
<code>gene_mean_celltype</code> represents the mean protein expression
for each sub-cell type specific to a spatial region. It is a matrix
consisting of 77 individuals and 4180 features. It is important to
acknowledge that within the context of our IMC data, the term “gene” is
used to refer to “protein”.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract cell type specific gene expression for all regions. </span></span>
<span><span class="va">gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">gene_mean_celltype</span></span>
<span></span>
<span><span class="co"># Extract HR+ CK7 cell type specific gene expression for Region5</span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span> <span class="st">"HR+ CK7--region5"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gene_mean_celltype</span><span class="op">)</span> , fixed<span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="va">gene_mean_celltype</span> <span class="op">[</span>, <span class="va">index</span><span class="op">]</span> </span>
<span></span>
<span><span class="co"># transpose to ensure we have gene by sample matrix</span></span>
<span><span class="va">gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">gene_mean_celltype</span><span class="op">)</span></span>
<span>      </span>
<span><span class="co"># Extract the two conditions of interest - poor prognosis vs good prognosis</span></span>
<span><span class="va">condition</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gene_mean_celltype</span><span class="op">)</span> , <span class="st">"_cond_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gene_mean_celltype</span><span class="op">)</span>, condition <span class="op">=</span> <span class="va">condition</span> <span class="op">)</span></span>
<span><span class="va">select_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span> <span class="va">condition</span><span class="op">$</span><span class="va">condition</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"poor"</span>,  <span class="st">"good"</span> <span class="op">)</span><span class="op">)</span></span>
<span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="va">condition</span><span class="op">[</span> <span class="va">select_index</span>, <span class="op">]</span></span>
<span><span class="va">gene_mean_celltype</span><span class="op">&lt;-</span> <span class="va">gene_mean_celltype</span> <span class="op">[</span> ,  <span class="va">select_index</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="co"># Calculate log fold change each protein using limma</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">condition</span>, data <span class="op">=</span> <span class="va">condition</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">gene_mean_celltype</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span><span class="va">tT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">fit</span>, n <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span>
<span><span class="va">tT</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tT</span><span class="op">)</span></span>
<span><span class="va">tT</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                                 logFC   AveExpr         t    P.Value adj.P.Val</span></span>
<span><span class="co">## HR+ CK7--region5--c_Myc    -0.2026693 0.3930301 -2.534247 0.01502026 0.2743783</span></span>
<span><span class="co">## HR+ CK7--region5--CAIX     -0.1398111 0.4082151 -2.202695 0.03306478 0.2743783</span></span>
<span><span class="co">## HR+ CK7--region5--Twist    -0.2007777 0.4171795 -2.170854 0.03554873 0.2743783</span></span>
<span><span class="co">## HR+ CK7--region5--vWF_CD31 -0.2401160 0.7328167 -2.036814 0.04789991 0.2743783</span></span>
<span><span class="co">## HR+ CK7--region5--EpCAM    -0.1988532 0.6477055 -2.031460 0.04846277 0.2743783</span></span>
<span><span class="co">##                                    B                       gene</span></span>
<span><span class="co">## HR+ CK7--region5--c_Myc    -3.168084    HR+ CK7--region5--c_Myc</span></span>
<span><span class="co">## HR+ CK7--region5--CAIX     -3.614202     HR+ CK7--region5--CAIX</span></span>
<span><span class="co">## HR+ CK7--region5--Twist    -3.654733    HR+ CK7--region5--Twist</span></span>
<span><span class="co">## HR+ CK7--region5--vWF_CD31 -3.820559 HR+ CK7--region5--vWF_CD31</span></span>
<span><span class="co">## HR+ CK7--region5--EpCAM    -3.827018    HR+ CK7--region5--EpCAM</span></span></code></pre>
<p>We visualise the comparison using a volcano plot and a dotplot for
the cell type specific expression feature. This is a type of
scatter-plot that is used to quickly identify changes in large datasets
and represent the significance (y-axis) versus effect size or
fold-change (x-axis).</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The volcano plot - Significance vs Effect Size</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span> <span class="va">tT</span> , <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">logFC</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">P.Value</span><span class="op">)</span> , text <span class="op">=</span> <span class="va">gene</span> <span class="op">)</span> <span class="op">)</span><span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">P.Value</span><span class="op">)</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, size<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"blue"</span>,high<span class="op">=</span><span class="st">"red"</span><span class="op">)</span><span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"log2 fold change"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"-log10 p-value"</span><span class="op">)</span>  </span>
<span></span>
<span><span class="co"># order the proteins by log fold change </span></span>
<span><span class="va">tT</span> <span class="op">&lt;-</span> <span class="va">tT</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">tT</span><span class="op">$</span><span class="va">logFC</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">tT</span> <span class="op">&lt;-</span> <span class="va">tT</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="op">]</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span> <span class="va">tT</span> , <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span> y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">gene</span>, <span class="va">logFC</span><span class="op">)</span> , x <span class="op">=</span> <span class="va">logFC</span>  <span class="op">)</span> <span class="op">)</span><span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">P.Value</span><span class="op">)</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, size<span class="op">=</span><span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"blue"</span>,high<span class="op">=</span><span class="st">"red"</span><span class="op">)</span><span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"logFC"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"region specific cell type specfic features"</span> <span class="op">)</span> </span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>,<span class="va">p2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-25-1.png" width="672"></p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Interactive Q&amp;A:</strong></p>
<p>Q4: Which figure do you prefer? The volcano plot or the dotplot?</p>
</div>
</div>
<div class="section level4">
<h4 class="tabset" id="can-we-identify-the-feature-of-interest-among-each-type-of-feature-category">Can we identify the feature of interest among each
type of feature category?<a class="anchor" aria-label="anchor" href="#can-we-identify-the-feature-of-interest-among-each-type-of-feature-category"></a>
</h4>
<p>In order to further develop our understanding of our data, it is
important to bear in mind that the outcomes derived from
<code>scFeatures</code> combine several association studies. Hence, it
is necessary to examine every separate feature category as an individual
unit of study. This section presents a graphic representation showing
how to visualise the distribution of features across individuals. The
following code has been developed to identify and retrieve all samples
that have been associated to both good and poor prognosis. This
particular step is necessary in this setting as our focus is limited to
two specific categories of individuals rather than including all
samples.</p>
<pre><code><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span></code></pre>
<div class="section level5">
<h5 class="unnumbered" id="gene-mean-celltype">Gene mean celltype<a class="anchor" aria-label="anchor" href="#gene-mean-celltype"></a>
</h5>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">gene_mean_celltype</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span><span class="co"># Select the features from HR+ CK7--cell type and in region 5</span></span>
<span><span class="va">feature</span>   <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span>  , <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span>  <span class="st">"HR+ CK7--region5"</span> , <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> , fixed<span class="op">=</span><span class="cn">T</span><span class="op">)</span> <span class="op">]</span> </span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Protein mean sub-cell type feature"</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="gene-mean-bulk">Gene mean bulk<a class="anchor" aria-label="anchor" href="#gene-mean-bulk"></a>
</h5>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">gene_mean_bulk</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Protein mean bulk feature"</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="nearest-neighbour-correlation">Nearest neighbour correlation<a class="anchor" aria-label="anchor" href="#nearest-neighbour-correlation"></a>
</h5>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">nn_correlation</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Nearest neighbour correlation feature"</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
</div>
</div>
<div class="section level4">
<h4 id="association-study-report">Association study report<a class="anchor" aria-label="anchor" href="#association-study-report"></a>
</h4>
<p>To accommodate for easier interpretation of the features,
<code>scFeatures</code> contains a function
<code>run_association_study_report</code> that enables the user to
readily visualise and explore all generated features with one line of
code. This should be used with caution as depending on the number of
cell-types considered, some of the automatically-generated graphics may
not be the most efficient way for interpretation.</p>
<p>In the code below we perform the following tasks:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Specify a folder to store the html report. Using the function
<code><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd()</a></code> we specify that we will store it in the current
working directory.</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>A small <code>for loop</code> is created to select all the samples
associated with good and poor prognosis and generate a new
scfeatures_result. This step is only necessary if the comparison of
interest is a subset of all your data.</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>Run the function <code><a href="https://sydneybiox.github.io/scFeatures/reference/run_association_study_report.html" class="external-link">run_association_study_report()</a></code>
</li>
</ol></li>
</ul>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Step 1 </span></span>
<span><span class="va">output_folder</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Step 2</span></span>
<span><span class="va">scfeatures_result_new</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">13</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">thisfeature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">thisfeature</span> <span class="op">&lt;-</span> <span class="va">thisfeature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"poor|good"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">thisfeature</span> <span class="op">)</span><span class="op">)</span>,  <span class="op">]</span></span>
<span>  <span class="va">scfeatures_result_new</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">thisfeature</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## Step 3</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/scFeatures/reference/run_association_study_report.html" class="external-link">run_association_study_report</a></span><span class="op">(</span><span class="va">scfeatures_result_new</span>, <span class="va">output_folder</span> <span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="can-we-estimate-recurrence-risk">Can we estimate recurrence risk?<a class="anchor" aria-label="anchor" href="#can-we-estimate-recurrence-risk"></a>
</h2>
<p>Recurrence risk estimation is a fundamental concern in medical
research, particularly in the context of patient survival analysis. In
this section, we will estimate recurrence risk using the molecular
representation of individuals generated from <code>scFeatures</code> to
build a survival model. We will use classifyR to build the survival
model. The patient outcome is time-to-event, so, by default, ClassifyR
will use Cox proportional hazards ranking to choose a set of features
and also Cox proportional hazards to predict risk scores. We will also
demonstrate other available models in <code>ClassifyR</code>.</p>
<div class="section level3">
<h3 id="building-a-survival-model">Building a survival model<a class="anchor" aria-label="anchor" href="#building-a-survival-model"></a>
</h3>
<p>Recall in the previous section that we have stored the 13 matrices of
different feature types in a list. Instead of manually retrieving each
matrix from the list to build separate models, classifyR can directly
take a list of matrices as an input and run a repeated cross-validation
model on each matrix individually. Below, we run 5 repeats of 5-fold
cross-validation. A high score indicates prognosis of a worse outcome
than a lower risk score. Although we have provided the code below, to
save time, <strong>just load the prepared RDS file
<code>classifyr_result_IMC.rds</code></strong> and we will focus on the
interpretation in this workshop.</p>
<pre><code><span><span class="co"># We use the following variables:     </span></span>
<span><span class="co"># timeRFS: "Time to Recurrence-Free Survival." It is the time period until recurrence occurs.    </span></span>
<span><span class="co"># eventRFS: "Event in Recurrence-Free Survival."It indicates whether the event has occurred.    </span></span>
<span><span class="co"># Breast.Tumour.Laterality: Laterality of tumors, eg, whether the tumor is located in left or right.     </span></span>
<span><span class="co"># ER.Status: Whether the tumor is ER positive or ER negative.    </span></span>
<span><span class="co"># Inferred.Menopausal.State: of the patient.    </span></span>
<span><span class="co"># Grade: of the tumor.   </span></span>
<span><span class="co"># Size: of the tumor.   </span></span>
<span></span>
<span><span class="va">usefulFeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Breast.Tumour.Laterality"</span>, <span class="st">"ER.Status"</span>, <span class="st">"Inferred.Menopausal.State"</span>, <span class="st">"Grade"</span>, <span class="st">"Size"</span><span class="op">)</span></span>
<span><span class="va">nFeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clinical <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">scfeatures_result</span>, <span class="kw">function</span><span class="op">(</span><span class="va">metaFeature</span><span class="op">)</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">clinicalAndOmics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clinical <span class="op">=</span> <span class="va">clinical</span><span class="op">)</span>, <span class="va">scfeatures_result</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### generate classfyr result </span></span>
<span><span class="va">classifyr_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span><span class="va">clinicalAndOmics</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"timeRFS"</span>, <span class="st">"eventRFS"</span><span class="op">)</span>,</span>
<span>                    extraParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>prepare <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>useFeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clinical <span class="op">=</span> <span class="va">usefulFeatures</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                    nFeatures <span class="op">=</span> <span class="va">nFeatures</span>, nFolds <span class="op">=</span> <span class="fl">5</span>, nRepeats <span class="op">=</span> <span class="fl">5</span>, nCores <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">classifyr_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/biocasia_data/classifyr_result_IMC.rds"</span><span class="op">)</span></span></code></pre></div>
<p>Cox proportional hazards is a classical statistical method, as
opposed to machine learning methods like Random survival forest. These
machine learning methods can build remarkably complex relationships
between features, however their running time can be much longer than Cox
proportional hazards. We use feature selection to limit the number of
features considered to at most 100 per metafeature and to save time,
<strong>you can just load the prepared RDS file.</strong> We will
compare the predictive performance between these methods.</p>
<pre><code><span><span class="va">nFeatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clinical <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">metaFeature</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">metaFeature</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">survForestCV</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span><span class="va">clinicalAndOmics</span>, <span class="va">outcome</span>, nFeatures <span class="op">=</span> <span class="va">nFeatures</span>,</span>
<span>                classifier <span class="op">=</span> <span class="st">"randomForest"</span>,</span>
<span>                nFolds <span class="op">=</span> <span class="fl">5</span>, nRepeats <span class="op">=</span> <span class="fl">5</span>, nCores <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">survForestCV</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/biocasia_data/survForestCV.RDS"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualising-the-model-performance-of-individual-metafeatures">Visualising the model performance of individual metafeatures<a class="anchor" aria-label="anchor" href="#visualising-the-model-performance-of-individual-metafeatures"></a>
</h3>
<p>To examine the distribution of prognostic performance, use
<code>performancePlot</code>. Currently, the only metric for
time-to-event data is C-index and that will automatically be used
because the predictive model type is tracked inside of the result
objects.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Make axis label 45 degree to improve readiability </span></span>
<span><span class="va">tilt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Putting two sets of cross-validation results together</span></span>
<span><span class="va">multiresults</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">classifyr_result</span>, <span class="va">survForestCV</span><span class="op">)</span></span>
<span><span class="va">ordering</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clinical"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html" class="external-link">performancePlot</a></span><span class="op">(</span><span class="va">multiresults</span>,</span>
<span>                characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span>, </span>
<span>                                           row <span class="op">=</span> <span class="st">"Classifier Name"</span><span class="op">)</span>,</span>
<span>                orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="va">ordering</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>                <span class="va">tilt</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
<p>Note how the resultant plot is a <code>ggplot2</code> object and can
be further modified. The same code could be used for a categorical
classifier because the random forest implementation provided by the
<code>ranger</code> package has the same interface for both. We will
examine feature selection stability with <code>selectionPlot</code>.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/selectionPlot.html" class="external-link">selectionPlot</a></span><span class="op">(</span><span class="va">multiresults</span>,</span>
<span>                characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span>, row <span class="op">=</span> <span class="st">"Classifier Name"</span><span class="op">)</span>,</span>
<span>                orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="va">ordering</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="va">tilt</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/distribution.html" class="external-link">distribution</a></span><span class="op">(</span><span class="va">classifyr_result</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      assay                   feature proportion</span></span>
<span><span class="co">## 1 clinical Inferred.Menopausal.State          1</span></span></code></pre>
<p>Using <code>samplesMetricMap</code> compare the per-sample C-index
for Cox models for all kinds of metafeatures.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/samplesMetricMap.html" class="external-link">samplesMetricMap</a></span><span class="op">(</span><span class="va">classifyr_result</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
<pre><code><span><span class="co">## TableGrob (2 x 1) "arrange": 2 grobs</span></span>
<span><span class="co">##   z     cells    name                 grob</span></span>
<span><span class="co">## 1 1 (2-2,1-1) arrange       gtable[layout]</span></span>
<span><span class="co">## 2 2 (1-1,1-1) arrange text[GRID.text.6176]</span></span></code></pre>
<p>A few samples are predicted better by some metafeatures than
others.</p>
<p>The per-sample C-index is a metric unique to ClassifyR. Models and
feature selection approaches may be seen in the vignette or listed by
<code><a href="https://sydneybiox.github.io/ClassifyR/reference/available.html" class="external-link">available()</a></code>.</p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i>
<i class="far fa-chart-bar fa-stack-1x fa-inverse"></i> </span>
<strong>Interactive Q&amp;A:</strong></p>
<p>Q7: Is the highest predictive performance the only way to choose the
best model or can other models be better for other reasons?</p>
</div>
</div>
<div class="section level3">
<h3 id="does-each-individual-require-a-different-collection-of-features">Does each individual require a different collection of
features?<a class="anchor" aria-label="anchor" href="#does-each-individual-require-a-different-collection-of-features"></a>
</h3>
<p>Using <code>samplesMetricMap</code> compare the per-sample C-index
for Cox models for all kinds of metafeatures.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/samplesMetricMap.html" class="external-link">samplesMetricMap</a></span><span class="op">(</span><span class="va">classifyr_result</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<pre><code><span><span class="co">## TableGrob (2 x 1) "arrange": 2 grobs</span></span>
<span><span class="co">##   z     cells    name                 grob</span></span>
<span><span class="co">## 1 1 (2-2,1-1) arrange       gtable[layout]</span></span>
<span><span class="co">## 2 2 (1-1,1-1) arrange text[GRID.text.6373]</span></span></code></pre>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i>
<i class="far fa-chart-bar fa-stack-1x fa-inverse"></i> </span>
<strong>Interactive Q&amp;A:</strong></p>
<p>Q8: Are spatial features important for predicting recurrence? Does it
hold for all individuals?</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<div class="section level3">
<h3 id="explanation-of-spatial-features">Explanation of spatial features<a class="anchor" aria-label="anchor" href="#explanation-of-spatial-features"></a>
</h3>
<ul>
<li>L function:</li>
</ul>
<p>The L function is a spatial statistic used to assess the spatial
distribution of cell types. It assesses the significance of cell-cell
interactions, by calculating the density of a cell type with other cell
types within a certain radius. High values indicate spatial association,
low values indicate spatial avoidance.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">one_sample</span>  <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span> , <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0128"</span>  <span class="op">]</span></span>
<span><span class="va">one_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">one_sample</span><span class="op">)</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">one_sample</span><span class="op">$</span><span class="va">description</span></span>
<span></span>
<span><span class="co"># select certain cell types to examine the interaction </span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span>  <span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span>  <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cells"</span>, <span class="st">"Fibroblasts"</span><span class="op">)</span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span><span class="op">!</span><span class="va">index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"others"</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span> <span class="va">one_sample</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span> <span class="va">celltype</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span> <span class="st">"Patient MB-0128 - high L value with \n B cells interacting Fibroblasts"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"coral"</span> , <span class="st">"steelblue"</span> , <span class="st">"grey70"</span><span class="op">)</span><span class="op">)</span></span>
<span> </span>
<span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">one_sample</span><span class="op">$</span><span class="va">description</span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span>  <span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span>  <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cells"</span>, <span class="st">"HR- CK7-"</span><span class="op">)</span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span><span class="op">!</span><span class="va">index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"others"</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span> <span class="va">one_sample</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span> <span class="va">celltype</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span>    <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span> <span class="st">"Patient MB-0128 - low L value with \n B cells interacting HR_ CK7"</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"coral"</span> , <span class="st">"steelblue"</span> , <span class="st">"grey70"</span><span class="op">)</span><span class="op">)</span></span>
<span> </span>
<span> </span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-36-1.png" width="960"></p>
<ul>
<li>Cell type interaction composition:</li>
</ul>
<p>We calculate the nearest neighbours of each cell and then calculate
the pairs of cell types based on the nearest neighbour. This allows us
to summarise it into a cell type interaction composition.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">one_sample</span>  <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span> , <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0263"</span>  <span class="op">]</span></span>
<span><span class="va">one_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">one_sample</span><span class="op">)</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">one_sample</span><span class="op">$</span><span class="va">description</span></span>
<span> </span>
<span><span class="va">a</span> <span class="op">&lt;-</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span> <span class="va">one_sample</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span> <span class="va">celltype</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/scale_color_tableau.html" class="external-link">scale_colour_tableau</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span> <span class="st">"Patient MB-0263"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">feature</span>  <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">celltype_interaction</span></span>
<span><span class="va">to_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span> <span class="va">feature</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"MB-0263"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span> , <span class="op">]</span><span class="op">)</span>  <span class="op">)</span></span>
<span><span class="va">to_plot</span><span class="op">$</span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">to_plot</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"value"</span>, <span class="st">"celltype interaction composition"</span><span class="op">)</span></span>
<span> </span>
<span><span class="va">to_plot</span> <span class="op">&lt;-</span> <span class="va">to_plot</span><span class="op">[</span> <span class="va">to_plot</span><span class="op">$</span><span class="va">value</span> <span class="op">&gt;</span> <span class="fl">0.03</span> , <span class="op">]</span> </span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">to_plot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">`celltype interaction composition`</span>, <span class="va">value</span><span class="op">)</span> ,  y <span class="op">=</span> <span class="va">value</span>, fill<span class="op">=</span><span class="va">`celltype interaction composition`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span> <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Major cell type interactions"</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-37-1.png" width="960"></p>
<ul>
<li>Moran’s I:</li>
</ul>
<p>Moran’s I is a spatial autocorrelation statistic based on both
location and values. It quantifies whether similar values tend to occur
near each other or are dispersed.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">high</span>  <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span><span class="st">"Ki67"</span>, <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0132"</span>  <span class="op">]</span></span>
<span><span class="va">high_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">high</span><span class="op">)</span> <span class="op">)</span> </span>
<span><span class="va">high_meta</span><span class="op">$</span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span> <span class="va">high</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">low</span>  <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span><span class="st">"Ki67"</span>,  <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span> <span class="st">"MB-0249"</span> <span class="op">]</span></span>
<span><span class="va">low_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">low</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">low_meta</span><span class="op">$</span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">low</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">high_meta</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Patient MB-0132 - high Moran's I in Ki67"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">low_meta</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Location_Center_X</span> , y <span class="op">=</span> <span class="va">Location_Center_Y</span>, colour <span class="op">=</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Patient MB-0249 - low Moran's I in Ki67"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-38-1.png" width="960"></p>
<ul>
<li>Nearest Neighbor Correlation:</li>
</ul>
<p>This metric measures the correlation of proteins/genes between a cell
and its nearest neighbour cell.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">plot_nncorrelation</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">thissample</span> , <span class="va">thisprotein</span><span class="op">)</span><span class="op">{</span></span>
<span>   </span>
<span>       <span class="va">sample_name</span> <span class="op">&lt;-</span> <span class="va">thissample</span></span>
<span>       <span class="va">thissample</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span>, <span class="va">data_sce</span><span class="op">$</span><span class="va">metabricId</span> <span class="op">==</span>     <span class="va">sample_name</span><span class="op">]</span></span>
<span>    </span>
<span>      </span>
<span>      <span class="va">exprsMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">)</span></span>
<span>     </span>
<span>    <span class="co"># calculate NN correlation </span></span>
<span>    <span class="va">cell_points_cts</span> <span class="op">&lt;-</span> <span class="fu">spatstat.geom</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spatstat.geom/man/ppp.html" class="external-link">ppp</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_X</span> <span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span>,</span>
<span>            check <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            xrange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_X</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_X</span><span class="op">)</span><span class="op">)</span></span>
<span>            <span class="op">)</span>,</span>
<span>            yrange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">thissample</span><span class="op">$</span><span class="va">Location_Center_Y</span><span class="op">)</span><span class="op">)</span></span>
<span>            <span class="op">)</span>,</span>
<span>            marks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">exprsMat</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>    </span>
<span>     <span class="va">value</span> <span class="op">&lt;-</span>  <span class="fu">spatstat.explore</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spatstat.explore/man/nncorr.html" class="external-link">nncorr</a></span><span class="op">(</span><span class="va">cell_points_cts</span><span class="op">)</span><span class="op">[</span><span class="st">"correlation"</span>, <span class="op">]</span></span>
<span>      <span class="va">value</span> <span class="op">&lt;-</span>  <span class="va">value</span><span class="op">[</span>  <span class="va">thisprotein</span><span class="op">]</span></span>
<span>     </span>
<span>    <span class="co"># Find the indices of the two nearest neighbors for each cell</span></span>
<span>    <span class="va">nn_indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spatstat.geom/man/nnwhich.html" class="external-link">nnwhich</a></span><span class="op">(</span><span class="va">cell_points_cts</span>, k <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">protein</span> <span class="op">&lt;-</span>  <span class="va">thisprotein</span></span>
<span>    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>thiscell_exprs  <span class="op">=</span> <span class="va">exprsMat</span><span class="op">[</span><span class="va">protein</span>, <span class="op">]</span> , exprs <span class="op">=</span>  <span class="va">exprsMat</span><span class="op">[</span><span class="va">protein</span>,<span class="va">nn_indices</span> <span class="op">]</span><span class="op">)</span></span>
<span>    </span>
<span>   <span class="va">p</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span> x <span class="op">=</span><span class="va">thiscell_exprs</span> ,  y <span class="op">=</span> <span class="va">exprs</span> , colour <span class="op">=</span>  <span class="va">exprs</span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span> <span class="st">"Patient "</span>, <span class="va">sample_name</span> ,  <span class="st">" nn_corr = "</span> ,  <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">value</span>, <span class="fl">2</span><span class="op">)</span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"This cell expression"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Neighbouring cell expression"</span><span class="op">)</span></span>
<span>   </span>
<span>   <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span> <span class="op">(</span><span class="va">p</span> <span class="op">)</span> </span>
<span></span>
<span><span class="op">}</span></span>
<span></span>
<span>    </span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">plot_nncorrelation</span><span class="op">(</span><span class="st">"MB-0605"</span>,  <span class="st">"HER2"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">plot_nncorrelation</span><span class="op">(</span><span class="st">"MB-0258"</span>,  <span class="st">"HER2"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="breastCancerIMC_files/figure-html/unnamed-chunk-39-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h3>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Debian GNU/Linux 12 (bookworm)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.21.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Australia/Sydney</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] reshape_0.8.9               scran_1.28.1               </span></span>
<span><span class="co">##  [3] scater_1.28.0               scuttle_1.10.1             </span></span>
<span><span class="co">##  [5] spatstat_3.0-6              spatstat.linnet_3.1-1      </span></span>
<span><span class="co">##  [7] spatstat.model_3.2-8        rpart_4.1.21               </span></span>
<span><span class="co">##  [9] spatstat.explore_3.2-5      nlme_3.1-163               </span></span>
<span><span class="co">## [11] spatstat.random_3.2-1       spatstat.geom_3.2-7        </span></span>
<span><span class="co">## [13] spatstat.data_3.0-3         survminer_0.4.9            </span></span>
<span><span class="co">## [15] ggpubr_0.6.0                tidyr_1.3.0                </span></span>
<span><span class="co">## [17] scattermore_1.2             plotly_4.10.3              </span></span>
<span><span class="co">## [19] limma_3.56.2                dplyr_1.1.3                </span></span>
<span><span class="co">## [21] spicyR_1.12.0               ggthemes_4.2.4             </span></span>
<span><span class="co">## [23] lisaClust_1.9.2             ClassifyR_3.5.21           </span></span>
<span><span class="co">## [25] survival_3.5-7              BiocParallel_1.34.2        </span></span>
<span><span class="co">## [27] MultiAssayExperiment_1.26.0 generics_0.1.3             </span></span>
<span><span class="co">## [29] scFeatures_0.99.27          ggplot2_3.4.4              </span></span>
<span><span class="co">## [31] SingleCellExperiment_1.22.0 SummarizedExperiment_1.30.2</span></span>
<span><span class="co">## [33] Biobase_2.60.0              GenomicRanges_1.52.0       </span></span>
<span><span class="co">## [35] GenomeInfoDb_1.36.1         IRanges_2.34.1             </span></span>
<span><span class="co">## [37] S4Vectors_0.38.1            BiocGenerics_0.46.0        </span></span>
<span><span class="co">## [39] MatrixGenerics_1.12.2       matrixStats_1.0.0          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.3                   ProtGenerics_1.32.0       </span></span>
<span><span class="co">##   [3] GSVA_1.48.2                spatstat.sparse_3.0-3     </span></span>
<span><span class="co">##   [5] bitops_1.0-7               httr_1.4.7                </span></span>
<span><span class="co">##   [7] RColorBrewer_1.1-3         numDeriv_2016.8-1.1       </span></span>
<span><span class="co">##   [9] EnsDb.Hsapiens.v79_2.99.0  tools_4.3.1               </span></span>
<span><span class="co">##  [11] backports_1.4.1            DT_0.30                   </span></span>
<span><span class="co">##  [13] utf8_1.2.4                 R6_2.5.1                  </span></span>
<span><span class="co">##  [15] HDF5Array_1.28.1           mgcv_1.9-0                </span></span>
<span><span class="co">##  [17] lazyeval_0.2.2             rhdf5filters_1.12.1       </span></span>
<span><span class="co">##  [19] withr_2.5.1                gridExtra_2.3             </span></span>
<span><span class="co">##  [21] prettyunits_1.2.0          cli_3.6.1                 </span></span>
<span><span class="co">##  [23] textshaping_0.3.7          labeling_0.4.3            </span></span>
<span><span class="co">##  [25] sass_0.4.7                 survMisc_0.5.6            </span></span>
<span><span class="co">##  [27] SingleCellSignalR_1.12.0   pkgdown_2.0.7             </span></span>
<span><span class="co">##  [29] Rsamtools_2.16.0           systemfonts_1.0.5         </span></span>
<span><span class="co">##  [31] ggupset_0.3.0              R.utils_2.12.2            </span></span>
<span><span class="co">##  [33] rstudioapi_0.15.0          RSQLite_2.3.1             </span></span>
<span><span class="co">##  [35] shape_1.4.6                BiocIO_1.10.0             </span></span>
<span><span class="co">##  [37] crosstalk_1.2.0            gtools_3.9.4              </span></span>
<span><span class="co">##  [39] car_3.1-2                  scam_1.2-14               </span></span>
<span><span class="co">##  [41] Matrix_1.6-1.1             ggbeeswarm_0.7.2          </span></span>
<span><span class="co">##  [43] fansi_1.0.5                abind_1.4-5               </span></span>
<span><span class="co">##  [45] R.methodsS3_1.8.2          lifecycle_1.0.3           </span></span>
<span><span class="co">##  [47] yaml_2.3.7                 edgeR_3.42.4              </span></span>
<span><span class="co">##  [49] carData_3.0-5              gplots_3.1.3              </span></span>
<span><span class="co">##  [51] rhdf5_2.44.0               BiocFileCache_2.8.0       </span></span>
<span><span class="co">##  [53] Rtsne_0.16                 blob_1.2.4                </span></span>
<span><span class="co">##  [55] promises_1.2.1             dqrng_0.3.1               </span></span>
<span><span class="co">##  [57] crayon_1.5.2               lattice_0.22-5            </span></span>
<span><span class="co">##  [59] cowplot_1.1.1              beachmat_2.16.0           </span></span>
<span><span class="co">##  [61] msigdbr_7.5.1              GenomicFeatures_1.52.1    </span></span>
<span><span class="co">##  [63] annotate_1.78.0            KEGGREST_1.40.0           </span></span>
<span><span class="co">##  [65] magick_2.8.1               pillar_1.9.0              </span></span>
<span><span class="co">##  [67] knitr_1.44                 metapod_1.8.0             </span></span>
<span><span class="co">##  [69] boot_1.3-28.1              rjson_0.2.21              </span></span>
<span><span class="co">##  [71] codetools_0.2-19           glue_1.6.2                </span></span>
<span><span class="co">##  [73] V8_4.4.0                   data.table_1.14.8         </span></span>
<span><span class="co">##  [75] vctrs_0.6.4                png_0.1-8                 </span></span>
<span><span class="co">##  [77] gtable_0.3.4               cachem_1.0.8              </span></span>
<span><span class="co">##  [79] xfun_0.40                  S4Arrays_1.0.6            </span></span>
<span><span class="co">##  [81] mime_0.12                  pheatmap_1.0.12           </span></span>
<span><span class="co">##  [83] iterators_1.0.14           KMsurv_0.1-5              </span></span>
<span><span class="co">##  [85] statmod_1.5.0              bluster_1.10.0            </span></span>
<span><span class="co">##  [87] ellipsis_0.3.2             bit64_4.0.5               </span></span>
<span><span class="co">##  [89] progress_1.2.2             filelock_1.0.2            </span></span>
<span><span class="co">##  [91] rprojroot_2.0.3            bslib_0.5.1               </span></span>
<span><span class="co">##  [93] irlba_2.3.5.1              vipor_0.4.5               </span></span>
<span><span class="co">##  [95] KernSmooth_2.23-22         colorspace_2.1-0          </span></span>
<span><span class="co">##  [97] DBI_1.1.3                  tidyselect_1.2.0          </span></span>
<span><span class="co">##  [99] proxyC_0.3.4               bit_4.0.5                 </span></span>
<span><span class="co">## [101] compiler_4.3.1             curl_5.0.2                </span></span>
<span><span class="co">## [103] AUCell_1.22.0              graph_1.78.0              </span></span>
<span><span class="co">## [105] BiocNeighbors_1.18.0       xml2_1.3.5                </span></span>
<span><span class="co">## [107] desc_1.4.2                 DelayedArray_0.26.6       </span></span>
<span><span class="co">## [109] rtracklayer_1.60.0         scales_1.2.1              </span></span>
<span><span class="co">## [111] caTools_1.18.2             rappdirs_0.3.3            </span></span>
<span><span class="co">## [113] SpatialExperiment_1.11.2   stringr_1.5.0             </span></span>
<span><span class="co">## [115] digest_0.6.33              goftest_1.2-3             </span></span>
<span><span class="co">## [117] minqa_1.2.6                spatstat.utils_3.0-4      </span></span>
<span><span class="co">## [119] rmarkdown_2.25             XVector_0.40.0            </span></span>
<span><span class="co">## [121] htmltools_0.5.6.1          pkgconfig_2.0.3           </span></span>
<span><span class="co">## [123] lme4_1.1-34                sparseMatrixStats_1.12.2  </span></span>
<span><span class="co">## [125] dbplyr_2.4.0               fastmap_1.1.1             </span></span>
<span><span class="co">## [127] ensembldb_2.24.0           htmlwidgets_1.6.2         </span></span>
<span><span class="co">## [129] rlang_1.1.1                GlobalOptions_0.1.2       </span></span>
<span><span class="co">## [131] shiny_1.7.5.1              DelayedMatrixStats_1.22.1 </span></span>
<span><span class="co">## [133] farver_2.1.1               jquerylib_0.1.4           </span></span>
<span><span class="co">## [135] zoo_1.8-12                 jsonlite_1.8.7            </span></span>
<span><span class="co">## [137] R.oo_1.25.0                BiocSingular_1.16.0       </span></span>
<span><span class="co">## [139] RCurl_1.98-1.12            magrittr_2.0.3            </span></span>
<span><span class="co">## [141] GenomeInfoDbData_1.2.10    Rhdf5lib_1.22.0           </span></span>
<span><span class="co">## [143] munsell_0.5.0              Rcpp_1.0.11               </span></span>
<span><span class="co">## [145] viridis_0.6.4              ape_5.7-1                 </span></span>
<span><span class="co">## [147] babelgene_22.9             stringi_1.7.12            </span></span>
<span><span class="co">## [149] zlibbioc_1.46.0            MASS_7.3-60               </span></span>
<span><span class="co">## [151] plyr_1.8.9                 ggrepel_0.9.4             </span></span>
<span><span class="co">## [153] parallel_4.3.1             deldir_1.0-9              </span></span>
<span><span class="co">## [155] Biostrings_2.68.1          splines_4.3.1             </span></span>
<span><span class="co">## [157] tensor_1.5                 multtest_2.56.0           </span></span>
<span><span class="co">## [159] hms_1.1.3                  circlize_0.4.15           </span></span>
<span><span class="co">## [161] locfit_1.5-9.8             igraph_1.5.1              </span></span>
<span><span class="co">## [163] ggsignif_0.6.4             reshape2_1.4.4            </span></span>
<span><span class="co">## [165] biomaRt_2.56.1             ScaledMatrix_1.8.1        </span></span>
<span><span class="co">## [167] XML_3.99-0.14              evaluate_0.22             </span></span>
<span><span class="co">## [169] RcppParallel_5.1.7         BiocManager_1.30.22       </span></span>
<span><span class="co">## [171] nloptr_2.0.3               tweenr_2.0.2              </span></span>
<span><span class="co">## [173] foreach_1.5.2              httpuv_1.6.12             </span></span>
<span><span class="co">## [175] EnsDb.Mmusculus.v79_2.99.0 purrr_1.0.2               </span></span>
<span><span class="co">## [177] polyclip_1.10-6            km.ci_0.5-6               </span></span>
<span><span class="co">## [179] ggforce_0.4.1              rsvd_1.0.5                </span></span>
<span><span class="co">## [181] broom_1.0.5                xtable_1.8-4              </span></span>
<span><span class="co">## [183] restfulr_0.0.15            AnnotationFilter_1.24.0   </span></span>
<span><span class="co">## [185] rstatix_0.7.2              later_1.3.1               </span></span>
<span><span class="co">## [187] viridisLite_0.4.2          class_7.3-22              </span></span>
<span><span class="co">## [189] ragg_1.2.6                 tibble_3.2.1              </span></span>
<span><span class="co">## [191] lmerTest_3.1-3             beeswarm_0.4.0            </span></span>
<span><span class="co">## [193] memoise_2.0.1              AnnotationDbi_1.62.2      </span></span>
<span><span class="co">## [195] GenomicAlignments_1.36.0   cluster_2.1.4             </span></span>
<span><span class="co">## [197] concaveman_1.1.0           GSEABase_1.62.0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="acknowledgment">Acknowledgment<a class="anchor" aria-label="anchor" href="#acknowledgment"></a>
</h3>
<p>The authors thank all their colleagues, particularly at The
University of Sydney, Sydney Precision Data Science and Charles Perkins
Centre for their support and intellectual engagement. Special thanks to
Ellis Patrick, Shila Ghazanfar, Andy Tran for guiding and supporting the
building of this workshop.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Sydney Precision Bioinformatics Alliance.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
