<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="BiocAsia2023">
<title>Unlocking single cell spatial omics analyses with scdney - Visium • BiocAsia2023</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Unlocking single cell spatial omics analyses with scdney - Visium">
<meta property="og:description" content="BiocAsia2023">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">BiocAsia2023</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/VisiumVersion3.html">Unlocking single cell spatial omics analyses with scdney - Visium</a>
    <a class="dropdown-item" href="../articles/breastCancerIMC.html">Unlocking single cell spatial omics analyses with scdney - IMC</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="VisiumVersion3_files/htmlwidgets-1.6.2/htmlwidgets.js"></script><link href="VisiumVersion3_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="VisiumVersion3_files/datatables-binding-0.30/datatables.js"></script><link href="VisiumVersion3_files/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="VisiumVersion3_files/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="VisiumVersion3_files/dt-core-1.13.4/js/jquery.dataTables.min.js"></script><link href="VisiumVersion3_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="VisiumVersion3_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Unlocking single cell spatial omics analyses with scdney - Visium</h1>
                        <h4 data-toc-skip class="author">Yue Cao, Andy
Tran, Dario Strbenac, Nicholas Robertson Jean Yang</h4>
            
            <h4 data-toc-skip class="date">15 October, 2023</h4>
      
      
      <div class="d-none name"><code>VisiumVersion3.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/scFeatures/" class="external-link">scFeatures</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/ClassifyR/" class="external-link">ClassifyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/lisaClust/" class="external-link">lisaClust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jrnold/ggthemes" class="external-link">ggthemes</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">spicyR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com" class="external-link">plotly</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scattermore</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival" class="external-link">survival</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/survminer/index.html" class="external-link">survminer</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://spatstat.org/" class="external-link">spatstat.geom</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarcElosua/SPOTlight" class="external-link">SPOTlight</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://had.co.nz/reshape" class="external-link">reshape</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># code for plotting purpose</span></span>
<span></span>
<span></span>
<span><span class="va">plot_boxplot</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span> <span class="va">feature</span> <span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">data_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span></span>
<span>  <span class="va">data_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape/man/melt-24.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">data_plot</span> <span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data_plot</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X1"</span>, <span class="st">"X2"</span>, <span class="st">"value"</span><span class="op">)</span></span>
<span>  <span class="va">data_plot</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span>  <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span> <span class="va">data_plot</span><span class="op">$</span><span class="va">X2</span><span class="op">)</span>, <span class="st">"_cond_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">data_plot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span> x <span class="op">=</span> <span class="va">X1</span>, y <span class="op">=</span> <span class="va">value</span> , colour <span class="op">=</span> <span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>   </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span>  </span>
<span></span>
<span><span class="va">plot_barplot</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">data</span> , <span class="va">dodge</span><span class="op">=</span><span class="cn">F</span>  <span class="op">)</span><span class="op">{</span></span>
<span></span>
<span>  <span class="va">data</span><span class="op">$</span><span class="va">patient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span> <span class="op">)</span>, <span class="st">"_cond_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">data</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span> <span class="op">)</span>, <span class="st">"_cond_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape/man/melt-24.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">data</span>, id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"patient"</span>, <span class="st">"condition"</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></span>
<span> </span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">data</span> , <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span> x <span class="op">=</span> <span class="va">patient</span> , y <span class="op">=</span> <span class="va">value</span> , fill <span class="op">=</span> <span class="va">variable</span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span>   </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span>   <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">condition</span>, scale<span class="op">=</span><span class="st">"free"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> </span>
<span> </span>
<span> <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span> <span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span> </span>
<span><span class="va">draw_dotplot</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">data_sce</span>, <span class="va">sample</span>, <span class="va">celltype</span> , <span class="va">group</span> , <span class="va">group_of_interest</span> <span class="op">)</span><span class="op">{</span></span>
<span></span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span> <span class="va">data_sce</span><span class="op">)</span><span class="op">)</span></span>
<span> </span>
<span>  <span class="va">df_plot</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>  <span class="va">thispatient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span> <span class="va">thispatient</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">this_df</span> <span class="op">&lt;-</span> <span class="va">df</span><span class="op">[</span><span class="va">df</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="va">thispatient</span>, <span class="op">]</span></span>
<span>    <span class="va">temp_df</span> <span class="op">&lt;-</span>   <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span> <span class="va">this_df</span><span class="op">$</span><span class="va">region</span>, <span class="va">this_df</span><span class="op">$</span><span class="va">celltype</span> <span class="op">)</span></span>
<span>    <span class="va">temp_df</span> <span class="op">&lt;-</span>  <span class="va">temp_df</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">temp_df</span><span class="op">)</span></span>
<span>    <span class="va">temp_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>  <span class="va">temp_df</span><span class="op">)</span></span>
<span>    <span class="va">temp_df</span><span class="op">$</span><span class="va">patient</span> <span class="op">&lt;-</span>  <span class="va">thispatient</span></span>
<span>    <span class="va">temp_df</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span> <span class="va">df</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span> <span class="va">df</span><span class="op">[[</span> <span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="va">thispatient</span> <span class="op">)</span>, <span class="op">]</span><span class="op">[[</span><span class="va">group</span><span class="op">]</span><span class="op">]</span> <span class="op">)</span> </span>
<span>    <span class="va">df_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">df_plot</span>, <span class="va">temp_df</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">df_plot</span> <span class="op">&lt;-</span> <span class="va">df_plot</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span> <span class="va">Var1</span> , <span class="va">Var2</span>, <span class="va">group</span> <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mean_proportion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Freq</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">df_plot</span>  <span class="op">&lt;-</span> <span class="va">df_plot</span> <span class="op">[</span> <span class="va">df_plot</span><span class="op">$</span><span class="va">group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">group_of_interest</span>, <span class="op">]</span></span>
<span>   </span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_plot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">Var2</span>, x <span class="op">=</span> <span class="va">Var1</span> ,colour <span class="op">=</span><span class="va">mean_proportion</span>  , size <span class="op">=</span> <span class="va">mean_proportion</span> <span class="op">)</span><span class="op">)</span><span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="op">~</span><span class="va">group</span>, scales <span class="op">=</span> <span class="st">"free"</span>, space <span class="op">=</span> <span class="st">"free"</span> <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Region"</span> <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Celltype"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="va">draw_region_clustering_result</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span> <span class="va">data_sce</span>, <span class="va">sample</span> , <span class="va">selected_sample</span>   <span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># get meta data for a selected patient to visualise </span></span>
<span>  <span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span></span>
<span>  <span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">metadata</span><span class="op">[</span><span class="va">metadata</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="va">selected_sample</span> ,  <span class="op">]</span></span>
<span>  <span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span>  <span class="va">metadata</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span> <span class="va">metadata</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">plotlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># define the list to store images for region highlighting </span></span>
<span>  <span class="va">plotlist_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># define the list to store images for celltype highlighting </span></span>
<span> </span>
<span>  <span class="va">thisregion</span>  <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">region</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  </span>
<span>  </span>
<span>  <span class="va">tableau_palette</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/scale_color_tableau.html" class="external-link">scale_colour_tableau</a></span><span class="op">(</span><span class="op">)</span> </span>
<span>  <span class="va">color_codes</span> <span class="op">&lt;-</span> <span class="va">tableau_palette</span><span class="op">$</span><span class="fu">palette</span><span class="op">(</span> <span class="fl">10</span> <span class="op">)</span></span>
<span>  <span class="va">color_codes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">color_codes</span>, <span class="st">"darkgrey"</span> , <span class="st">"grey90"</span><span class="op">)</span> </span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">color_codes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="va">celltype</span> ,  <span class="st">"other regions"</span><span class="op">)</span></span>
<span>  </span>
<span>  </span>
<span>  <span class="co"># put this in a function </span></span>
<span>  <span class="co"># show the hatching plot as well </span></span>
<span>  <span class="kw">for</span> <span class="op">(</span> <span class="va">thisregion</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">region</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>          </span>
<span>          <span class="va">selected_region_index</span> <span class="op">&lt;-</span>  <span class="va">metadata</span><span class="op">$</span><span class="va">region</span> <span class="op">==</span> <span class="va">thisregion</span></span>
<span>          </span>
<span>          <span class="va">metadata</span><span class="op">$</span><span class="va">selected_region</span> <span class="op">&lt;-</span>  <span class="st">"other regions"</span></span>
<span>          <span class="va">metadata</span><span class="op">$</span><span class="va">selected_region</span><span class="op">[</span><span class="va">selected_region_index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"selected region"</span></span>
<span>          </span>
<span>          <span class="va">metadata</span><span class="op">$</span><span class="va">selected_celltype</span> <span class="op">&lt;-</span> <span class="va">metadata</span><span class="op">$</span><span class="va">celltype</span></span>
<span>          <span class="va">metadata</span><span class="op">$</span><span class="va">selected_celltype</span><span class="op">[</span><span class="op">!</span><span class="va">selected_region_index</span> <span class="op">]</span> <span class="op">&lt;-</span>   <span class="st">"other regions"</span></span>
<span>          </span>
<span>          <span class="co"># metadata$celltype &lt;- factor(metadata$celltype, levels = c(unique(metadata$celltype), "other regions"))</span></span>
<span>  </span>
<span>         <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">spatial_x</span> , y <span class="op">=</span> <span class="va">spatial_y</span>, colour <span class="op">=</span> <span class="va">selected_region</span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>           <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span> alpha <span class="op">=</span> <span class="fl">0.8</span> <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">thisregion</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span> , <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span>           </span>
<span>         </span>
<span>      </span>
<span>         <span class="va">p2</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">spatial_x</span> , y <span class="op">=</span> <span class="va">spatial_y</span>, colour <span class="op">=</span>  <span class="va">selected_celltype</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>           <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.8</span> <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">thisregion</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span>  <span class="va">color_codes</span><span class="op">)</span></span>
<span>         </span>
<span>        <span class="va">plotlist</span> <span class="op">[[</span><span class="va">thisregion</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">p</span></span>
<span>         </span>
<span>        <span class="va">plotlist_celltype</span> <span class="op">[[</span><span class="va">thisregion</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">p2</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>   <span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plotlist</span> , ncol <span class="op">=</span> <span class="fl">5</span>, nrow <span class="op">=</span> <span class="fl">1</span> , common.legend <span class="op">=</span> <span class="cn">T</span> <span class="op">)</span></span>
<span>   <span class="va">b</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plotlist_celltype</span> , ncol <span class="op">=</span> <span class="fl">5</span>, nrow <span class="op">=</span> <span class="fl">1</span> , common.legend <span class="op">=</span> <span class="cn">T</span> <span class="op">)</span></span>
<span>  </span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="va">draw_selected_region_boxplot</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">data_sce</span>, <span class="va">sample</span> , <span class="va">celltype</span>, <span class="va">group</span> ,  <span class="va">group_of_interest</span>, <span class="va">select_region</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>   <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span> <span class="va">data_sce</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    </span>
<span>    </span>
<span>    <span class="va">df_plot</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span> <span class="va">thispatient</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="va">this_df</span> <span class="op">&lt;-</span> <span class="va">df</span><span class="op">[</span><span class="va">df</span><span class="op">[[</span><span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="va">thispatient</span>, <span class="op">]</span></span>
<span>      <span class="va">temp_df</span> <span class="op">&lt;-</span>   <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span> <span class="va">this_df</span><span class="op">$</span><span class="va">region</span>, <span class="va">this_df</span><span class="op">[[</span><span class="va">celltype</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>      <span class="va">temp_df</span> <span class="op">&lt;-</span>  <span class="va">temp_df</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">temp_df</span><span class="op">)</span></span>
<span>      <span class="va">temp_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>  <span class="va">temp_df</span><span class="op">)</span></span>
<span>      <span class="va">temp_df</span><span class="op">$</span><span class="va">patient</span> <span class="op">&lt;-</span>  <span class="va">thispatient</span></span>
<span>      <span class="va">temp_df</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span> <span class="va">df</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span> <span class="va">df</span><span class="op">[[</span> <span class="va">sample</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="va">thispatient</span> <span class="op">)</span>, <span class="op">]</span><span class="op">[[</span><span class="va">group</span><span class="op">]</span><span class="op">]</span> <span class="op">)</span> </span>
<span>      <span class="va">df_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">df_plot</span>, <span class="va">temp_df</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="va">df_plot_region_1</span> <span class="op">&lt;-</span> <span class="va">df_plot</span><span class="op">[</span><span class="va">df_plot</span><span class="op">$</span><span class="va">Var1</span> <span class="op">==</span> <span class="va">select_region</span>, <span class="op">]</span></span>
<span>   </span>
<span>    <span class="va">df_plot_region_1</span>   <span class="op">&lt;-</span>  <span class="va">df_plot_region_1</span>  <span class="op">[</span>  <span class="va">df_plot_region_1</span><span class="op">$</span><span class="va">group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">group_of_interest</span>, <span class="op">]</span></span>
<span>    </span>
<span>    <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_plot_region_1</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span>  <span class="va">Var2</span>,  y <span class="op">=</span> <span class="va">Freq</span>, colour <span class="op">=</span> <span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span> </span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Proportion"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Cell type"</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Region 2"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span>
<span>      </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">p</span> <span class="op">)</span></span>
<span> </span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>As single cell technology advances, the recent development of spatial
omics allows us to examine the spatial organisation of cells within
tissues in their native environment. This case study will discuss the
challenges and analytical focus associated with using multi-sample
spatial datasets for disease risk prognosis. We will also talk about
general analytical strategies and the critical thinking questions that
arise in the workflow.</p>
<br><div class="aimbox">
<div class="section level4">
<h4 class="unnumbered" id="preparation-and-assumed-knowledge">
<span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-pencil-alt fa-stack-1x fa-inverse"></i></span>
Preparation and assumed knowledge<a class="anchor" aria-label="anchor" href="#preparation-and-assumed-knowledge"></a>
</h4>
<ul>
<li>Knowledge of R syntax</li>
<li>Familiarity with the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment
class</a>
</li>
<li>Basic knowledge in <a href="https://bioconductor.org/books/release/OSCA/index.html" class="external-link">single
cell data analysis</a>. You can access our <a href="https://github.com/SydneyBioX/scdney#scdney-workshops-series" class="external-link">previous
workshops</a> for a quick review in single cell data analysis.</li>
<li>Ability to install all required R packages, please check
<code>sessionInfo</code> at the end of this document to ensure you are
using the correct version.</li>
<li>Familiarity with our previous workshop vignette on <a href="https://sydneybiox.github.io/BIS2019_SC/" class="external-link">Introduction to Single
Cell RNA-seq Analysis</a>
</li>
</ul>
</div>
<div class="section level4">
<h4 class="unnumbered" id="learning-objectives">
<span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-location-arrow fa-stack-1x fa-inverse"></i></span>
Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h4>
<ul>
<li>Describe and visualise spatial omics datasets<br>
</li>
<li>Calculate different measures that describe the spatial distribution
of cell types</li>
<li>Generate individual feature representations from a cell-level
expression matrix</li>
<li>Perform multi-view disease outcome prognosis with the package
<code>ClassifyR</code>
</li>
<li>Develop understanding on
<ul>
<li>how to assess the performance of classification and survival
models<br>
</li>
<li>how to identify and assess individual performance given a survival
model.</li>
</ul>
</li>
<li>Explore various strategies for disease outcome prognosis using
spatial omics data<br>
</li>
</ul>
</div>
</div>
<p><strong>Note:</strong> This data analysis workshop offers
participants the opportunity to engage in hands-on analysis using R.
However, if you are not comfortable with coding in R, you can still
acquire valuable interpretation skills by reviewing the output we
provide in this file. <br></p>
<div class="section level4">
<h4 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h4>
<p>Structure of the 3-hour workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Exploring spatial data</td>
<td>35m</td>
</tr>
<tr class="odd">
<td>Feature engineering with scFeatures</td>
<td>50m</td>
</tr>
<tr class="even">
<td>Break</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Survival analysis with ClassifyR</td>
<td>45m</td>
</tr>
<tr class="even">
<td>Identify cohort heterogeneity</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Concluding remarks</td>
<td></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="initial-exploration-and-visualisation">Initial exploration and visualisation<a class="anchor" aria-label="anchor" href="#initial-exploration-and-visualisation"></a>
</h2>
<div class="section level5">
<h5 class="unnumbered" id="data-and-background">Data and background<a class="anchor" aria-label="anchor" href="#data-and-background"></a>
</h5>
<p>In this demo, we look at a Visium dataset taken from Kuppe, C.,
Ramirez Flores, R. O., Li, Z., Hayat, S., Levinson, R. T., Liao, X., …
&amp; Kramann, R. (2022). Spatial multi-omic map of human myocardial
infarction. Nature, 608(7924), 766-777.</p>
<p>Visium captures spatial information, creating images that display the
distribution of different cell types and their associated gene
expression patterns in the tissue.<br>
In this dataset, the authors quantified the expression of &gt;10000
genes in control and in patients with myocardial infarction. In this
demo, we examine patients defined to be in the myogenic group and
ischaemic group. The myogenic group is defined by samples taken from
control and the unaffected remote zone, and the ischaemic group is
defined by samples taken from the ischaemic zone.</p>
</div>
<div class="section level3">
<h3 id="exploration-1-how-complex-is-my-data">Exploration 1: How complex is my data ?<a class="anchor" aria-label="anchor" href="#exploration-1-how-complex-is-my-data"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/biocasia_data/small_data.rds"</span><span class="op">)</span></span>
<span> </span>
<span><span class="va">data_sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 11681 19000 </span></span>
<span><span class="co">## metadata(3240): X_approximate_distribution X_normalization ...</span></span>
<span><span class="co">##   schema_version title</span></span>
<span><span class="co">## assays(2): X logcounts</span></span>
<span><span class="co">## rownames(11681): LINC01409 LINC01128 ... VAMP7 AC007325.4</span></span>
<span><span class="co">## rowData names(1): features</span></span>
<span><span class="co">## colnames(19000): ACAGAACTGAGAACAA-1 TTGCCCTGATCACGGG-1 ...</span></span>
<span><span class="co">##   CTGACATAGAAATAGA-1.16 ATTAGGCGATGCTTTC-1.24</span></span>
<span><span class="co">## colData names(33): n_counts n_genes ... number_cells condition</span></span>
<span><span class="co">## reducedDimNames(3): X_pca X_spatial X_umap</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(0) :</span></span>
<span><span class="co">## imgData names(1): sample_id</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Expression matrix is stored in genes by cells matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 7 x 7 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##           ACAGAACTGAGAACAA-1 TTGCCCTGATCACGGG-1 CATATAGGTACAGTCA-1</span></span>
<span><span class="co">## LINC01409           .                  .                  .       </span></span>
<span><span class="co">## LINC01128           1.431438           .                  1.351835</span></span>
<span><span class="co">## NOC2L               .                  .                  .       </span></span>
<span><span class="co">## KLHL17              .                  .                  .       </span></span>
<span><span class="co">## HES4                .                  .                  .       </span></span>
<span><span class="co">## ISG15               .                  1.279462           .       </span></span>
<span><span class="co">## AGRN                .                  .                  .       </span></span>
<span><span class="co">##           TGCGAGAATATTACCC-1 TATTCAATTCTAATCC-1 GTCCTATTGTTGTGGT-1</span></span>
<span><span class="co">## LINC01409           .                  .                  .       </span></span>
<span><span class="co">## LINC01128           .                  .                  .       </span></span>
<span><span class="co">## NOC2L               .                  .                  .       </span></span>
<span><span class="co">## KLHL17              .                  .                  .       </span></span>
<span><span class="co">## HES4                1.146986           1.389528           .       </span></span>
<span><span class="co">## ISG15               .                  .                  1.454652</span></span>
<span><span class="co">## AGRN                .                  .                  .       </span></span>
<span><span class="co">##           TAACCTACCGTCCGAG-1</span></span>
<span><span class="co">## LINC01409           .       </span></span>
<span><span class="co">## LINC01128           .       </span></span>
<span><span class="co">## NOC2L               .       </span></span>
<span><span class="co">## KLHL17              .       </span></span>
<span><span class="co">## HES4                .       </span></span>
<span><span class="co">## ISG15               1.475897</span></span>
<span><span class="co">## AGRN                .</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## The object stores meta data (such as patient outcome information) about each cell</span></span>
<span><span class="co">## the metadata is stored in colData() for a spatialexperiment object.</span></span>
<span><span class="co"># we need to convert to a data.frame to be input into datatable function</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span>, options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>scrollX <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-fe7ba4b1286cb2b4140c" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-fe7ba4b1286cb2b4140c">{"x":{"filter":"none","vertical":false,"data":[["ACAGAACTGAGAACAA-1","TTGCCCTGATCACGGG-1","CATATAGGTACAGTCA-1","TGCGAGAATATTACCC-1","TATTCAATTCTAATCC-1"],[3140,3854,3491,4654,3319],[1574,1966,1860,2202,1684],[29.07253269916766,25.02499166944352,27.24382946896036,28.32196452933152,29.90601503759398],[0.02770382138137779,0.01671721367317223,0.0004733077431533187,0.003617150797638009,0.01485492533975777],[0.5699686404745494,0.2373769981738169,0.3612975692095668,0.4256981750846177,0.6680740575036851],[0.07051242985528622,0.1531399472771346,0.01289209726431274,0.07994222649116965,0.03141037068305097],[0.1628007724292061,0.4126826103846002,0.2362456893131534,0.2435250410466519,0.1524150907414053],[0.004343729637012451,0.005865748702226788,0.003204839823714372,0.005206225283947605,0.01172116037138452],[0.0100938729062239,0.03401375574114668,0.002340662715472429,0.001832455813278091,0.002487665511826019],[0.0821746491735931,0.02471130606794164,0.3032342838578256,0.0287272085132965,0.02455097897393392],[0.004084990967738683,0.009724305844748399,0.01341718278564213,0.00307941815936564,0.04439387584216319],[0.04974388752547061,0.07706423433524595,0.03346700325611637,0.1122836109201789,0.01145328746286508],[0.00447534201512893,0.02056689618937275,0.02861482860924984,0.08978167835932485,0.02130161616394754],[0.01409786363441281,0.008136983610593758,0.004812535421792961,0.006306809530531224,0.01733697140598063],["Cardiomyocyte","Fibroblast","Cardiomyocyte","Cardiomyocyte","Cardiomyocyte"],["EFO:0010961","EFO:0010961","EFO:0010961","EFO:0010961","EFO:0010961"],["CL:0000513","CL:0002548","CL:0000513","CL:0000513","CL:0000513"],["HsapDv:0000138","HsapDv:0000138","HsapDv:0000138","HsapDv:0000138","HsapDv:0000138"],["PATO:0000461","PATO:0000461","PATO:0000461","PATO:0000461","PATO:0000461"],["HANCESTRO:0005","HANCESTRO:0005","HANCESTRO:0005","HANCESTRO:0005","HANCESTRO:0005"],[true,true,true,true,true],["NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606","NCBITaxon:9606"],["PATO:0000383","PATO:0000383","PATO:0000383","PATO:0000383","PATO:0000383"],["UBERON:0002084","UBERON:0002084","UBERON:0002084","UBERON:0002084","UBERON:0002084"],["sample01.1","sample01.1","sample01.1","sample01.1","sample01.1"],["control_P1","control_P1","control_P1","control_P1","control_P1"],["control","control","control","control","control"],[0.1889763779527559,0.2677165354330709,0.4330708661417323,0.3149606299212598,0.5826771653543307],[0.3246753246753247,0.5324675324675324,0.4415584415584415,0.3506493506493507,0.5584415584415584],["Cardiomyocyte","Fibroblast","Cardiomyocyte","Cardiomyocyte","Cardiomyocyte"],["control_P1_cond_Myogenic","control_P1_cond_Myogenic","control_P1_cond_Myogenic","control_P1_cond_Myogenic","control_P1_cond_Myogenic"],[68.95830119531992,75.95378349660785,74.96910334879338,77.88308444550651,71.03197625523696],["Myogenic","Myogenic","Myogenic","Myogenic","Myogenic"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>n_counts<\/th>\n      <th>n_genes<\/th>\n      <th>percent.mt<\/th>\n      <th>Adipocyte<\/th>\n      <th>Cardiomyocyte<\/th>\n      <th>Endothelial<\/th>\n      <th>Fibroblast<\/th>\n      <th>Lymphoid<\/th>\n      <th>Mast<\/th>\n      <th>Myeloid<\/th>\n      <th>Neuronal<\/th>\n      <th>Pericyte<\/th>\n      <th>Cycling.cells<\/th>\n      <th>vSMCs<\/th>\n      <th>cell_type_original<\/th>\n      <th>assay_ontology_term_id<\/th>\n      <th>cell_type_ontology_term_id<\/th>\n      <th>development_stage_ontology_term_id<\/th>\n      <th>disease_ontology_term_id<\/th>\n      <th>ethnicity_ontology_term_id<\/th>\n      <th>is_primary_data<\/th>\n      <th>organism_ontology_term_id<\/th>\n      <th>sex_ontology_term_id<\/th>\n      <th>tissue_ontology_term_id<\/th>\n      <th>sample_id<\/th>\n      <th>Patient<\/th>\n      <th>Type<\/th>\n      <th>spatial_x<\/th>\n      <th>spatial_y<\/th>\n      <th>celltype<\/th>\n      <th>sample<\/th>\n      <th>number_cells<\/th>\n      <th>condition<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,28,29,32]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script><p>The code above examine the data objects:<br>
- The dataset contains 11681 genes and 19000 cells (after
subsampling).<br>
- The outcome is 11 myogenic samples (4 from control, 7 from remote
zone) and 8 ischaemic samples.</p>
</div>
<div class="section level3">
<h3 id="exploration-2-how-to-visualise-my-data">Exploration 2: How to visualise my data?<a class="anchor" aria-label="anchor" href="#exploration-2-how-to-visualise-my-data"></a>
</h3>
<p>Typically in spatial data, we perform dimension reduction to reduce
and project the high dimensional cell by gene matrix onto 2D space. This
allows us to visualise various things of interest, such as the
distribution of cell types and disease outcomes.<br>
Visium is a spot-based technology, meaning that each spot captures 1-10
cells. Therefore each spot may represent a mixture of cells from
different cell types. In this dataset, the author used cell2location
which is a spatial deconvolution tool to predict the cell type
probability or composition of each of the spot. There are also many
other deconvolution tools available, eg, CARD published in Ma, Y., &amp;
Zhou, X. (2022). Spatially informed cell-type deconvolution for spatial
transcriptomics. Nature biotechnology, 40(9), 1349-1359.</p>
<p>For demonstration purposes, in the plot below we visualise each spot
using the cell type with maximum probability. Remember, the focus is to
identify any interesting patterns from the plot.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Perform UMP calculation</span></span>
<span><span class="va">data_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">data_sce</span>, exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>, scale<span class="op">=</span><span class="cn">TRUE</span>, min_dist <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Visualize UMAP and highlight variable of interest</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">data_sce</span>, colour_by <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">data_sce</span>, colour_by <span class="op">=</span> <span class="st">"condition"</span><span class="op">)</span></span>
<span><span class="va">c</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">data_sce</span>, colour_by <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Combine multiple plot in a grid format </span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i>
<i class="far fa-chart-bar fa-stack-1x fa-inverse"></i> </span>
<strong>Critical Thinking Q:</strong></p>
<ul>
<li>Q1: Is there patient batch effect?<br>
</li>
<li>Q2: Are the myogenic group and ischaemic group easy or difficult to
distinguish?<br>
</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 class="tabset" id="exploration-3-is-there-spatial-structure-in-my-data">Exploration 3: Is there spatial structure in my
data?<a class="anchor" aria-label="anchor" href="#exploration-3-is-there-spatial-structure-in-my-data"></a>
</h3>
<p>The advantage with spatial omics is that we can examine the
organisation of the cell types as it occurs on the tissue slide. Here,
we visualise one of the slides from a patient. As an optional exercise,
you can: - permute the cell type label, and<br>
- permute the spatial coordinate.<br>
to give a sense of what is random ordering.</p>
<p>The code below is to set a consistent colour for each cell type
throughout this rmarkdown. We use the colour palette Tableau because it
only has 10 colours and we have 11 cell types, and we colour the last
cell type grey.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SET UP code for visualising </span></span>
<span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Adipocyte"</span> , <span class="st">"Cardiomyocyte"</span> , </span>
<span>              <span class="st">"Endothelial"</span> ,  <span class="st">"Fibroblast"</span> , <span class="st">"Lymphoid"</span> ,    </span>
<span>              <span class="st">"Mast"</span> ,   <span class="st">"Myeloid"</span> , </span>
<span>              <span class="st">"Neuronal"</span> , <span class="st">"Pericyte"</span> , <span class="st">"Cycling.cells"</span>, <span class="st">"vSMCs"</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">tableau_palette</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/scale_color_tableau.html" class="external-link">scale_colour_tableau</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">color_codes</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="va">tableau_palette</span><span class="op">$</span><span class="fu">palette</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span> , <span class="st">"lightgrey"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">color_codes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">celltype</span></span></code></pre></div>
<div class="section level4">
<h4 class="unnumbered" id="spatial-plot">Spatial plot<a class="anchor" aria-label="anchor" href="#spatial-plot"></a>
</h4>
<p>We select a particular patient “IZ_P9_cond_Ischaemic” (a patient in
the ischaemic group) and visualise its spatial pattern using ggplot.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select one sample to visualise  </span></span>
<span><span class="va">one_sample_data</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">[</span>, <span class="va">data_sce</span><span class="op">$</span><span class="va">sample</span>  <span class="op">==</span>  <span class="st">"IZ_P9_cond_Ischaemic"</span><span class="op">]</span></span>
<span><span class="va">one_sample</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">one_sample_data</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">one_sample</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">spatial_x</span> , y <span class="op">=</span> <span class="va">spatial_y</span>, colour <span class="op">=</span> <span class="va">celltype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_codes</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Original slide"</span><span class="op">)</span></span>
<span><span class="va">a</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 class="unnumbered" id="optional-code-random-spatial-pattern">[Optional code] Random spatial pattern<a class="anchor" aria-label="anchor" href="#optional-code-random-spatial-pattern"></a>
</h4>
<p>The code here investigates permutations of spatially resolved data.
Please examine the next tab for actual results.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Optional: Permute the cell type labels</span></span>
<span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype_permute</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">one_sample</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">one_sample</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">spatial_x</span> , y <span class="op">=</span>  <span class="va">spatial_y</span>, colour <span class="op">=</span><span class="va">celltype_permute</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_codes</span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Permute the cell type label"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Optional: Permute the spatial coordinate</span></span>
<span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">spatial_x_permute</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">one_sample</span><span class="op">$</span><span class="va">spatial_x</span><span class="op">)</span></span>
<span><span class="va">one_sample</span><span class="op">$</span><span class="va">spatial_y_permute</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">one_sample</span><span class="op">$</span><span class="va">spatial_y</span><span class="op">)</span></span>
<span></span>
<span><span class="va">c</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">one_sample</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">spatial_x_permute</span> , y <span class="op">=</span> <span class="va">spatial_y_permute</span>, colour <span class="op">=</span> <span class="va">celltype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_codes</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Permute the X, Y coordinate"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 class="unnumbered" id="spatial-structure">Spatial structure<a class="anchor" aria-label="anchor" href="#spatial-structure"></a>
</h4>
<p>The aim here is to have an understanding of the concept of spatial
randomness. Spatial statistics is a topic of study that encompasses a
wide range of research. The next two code chucks focus on the
examination of two distinct permutation strategies. The objective of
this investigation is to gain an understanding of how various
permutation strategies might yield varied perceptions of randomness.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span>, ncol <span class="op">=</span><span class="fl">3</span> <span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-9-1.png" width="1152"></p>
</div>
</div>
<div class="section level3">
<h3 id="exploration-4-visium-specific-visualisation">Exploration 4: Visium specific visualisation<a class="anchor" aria-label="anchor" href="#exploration-4-visium-specific-visualisation"></a>
</h3>
<p>Instead of plotting the cell type with maximum probability, we can
also visualise the cell type composition of each spot using a pie
chart.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> imagecol <span class="op">=</span> <span class="va">one_sample</span><span class="op">$</span><span class="va">spatial_y</span>, </span>
<span>                 imagerow <span class="op">=</span> <span class="va">one_sample</span><span class="op">$</span><span class="va">spatial_x</span><span class="op">)</span></span>
<span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Spot"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">one_sample_data</span><span class="op">)</span><span class="op">[</span> ,<span class="va">celltype</span>  <span class="op">]</span> <span class="op">)</span></span>
<span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Spot"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## we use a function to plot the piechart</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SPOTlight/man/plotSpatialScatterpie.html" class="external-link">plotSpatialScatterpie</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span> , pie_scale <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_codes</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"spatial_y"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"spatial_y"</span><span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i>
<i class="far fa-chart-bar fa-stack-1x fa-inverse"></i> </span>
<strong>Critical Thinking Q:</strong></p>
<p>Q3: Is there structure in the data or is the cell type randomly
distribution?</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="describing-tissue-microenvrionments-and-cellular-neighbourhoods">Describing tissue microenvrionments and cellular neighbourhoods<a class="anchor" aria-label="anchor" href="#describing-tissue-microenvrionments-and-cellular-neighbourhoods"></a>
</h2>
<div class="section level3">
<h3 id="do-cell-type-co-localise-in-specfic-regions">Do cell type co-localise in specfic regions?<a class="anchor" aria-label="anchor" href="#do-cell-type-co-localise-in-specfic-regions"></a>
</h3>
<p>Spatial data allow for the identification of a variety of
characteristics, including distinct cell types within an image, which
provide an overview of the tissue environment. This allows scientists to
explore the cellular architecture and environment and its association
with phenotype information (e.g meta-data). For our data story, we are
interested in whether ischaemic and myogenic individuals are different.
In this section, we examine graphically how cell-type co-localisation
varies across spatial regions and how such information is associated
with disease conditions. There are many packages that perform this task.
Here, we use the lisaClust function [<a href="https://www.bioconductor.org/packages/devel/bioc/html/lisaClust.html" class="external-link uri">https://www.bioconductor.org/packages/devel/bioc/html/lisaClust.html</a>]
that is based on the “local L-function” to spatially cluster cells into
different regions with similar cell type compositions.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span> </span>
<span><span class="va">BPPARAM</span> <span class="op">&lt;-</span> <span class="fu">simpleSeg</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/pkg/simpleSeg/man/generateBPParam.html" class="external-link">generateBPParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Cluster cells into spatial regions with similar composition.</span></span>
<span><span class="va">data_sce</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/lisaClust.html" class="external-link">lisaClust</a></span><span class="op">(</span></span>
<span>  <span class="va">data_sce</span> ,</span>
<span>  k <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  Rs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>  sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"spatial_x"</span>, <span class="st">"spatial_y"</span><span class="op">)</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"celltype"</span>,</span>
<span>  imageID <span class="op">=</span> <span class="st">"sample"</span> ,</span>
<span>  regionName <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 class="tabset" id="which-regions-appear-to-be-different-between-the-ischaemic-and-myogenic-patients">Which regions appear to be different between the
ischaemic and myogenic patients?<a class="anchor" aria-label="anchor" href="#which-regions-appear-to-be-different-between-the-ischaemic-and-myogenic-patients"></a>
</h3>
<p>We have use the <code>lisaClust</code> function in the previous
subsection to cluster cells into five different spatial regions. Next,
as a case study, we will compare individuals with ischaemic and myogenic
disease and examine graphically, if any regions appear to be different
between disease conditions.</p>
<div class="section level4">
<h4 class="unnumbered" id="individual-level">Individual level<a class="anchor" aria-label="anchor" href="#individual-level"></a>
</h4>
<p>Here, we visualise the spatial domain (region) detection result based
on one individual. We can either visualise all regions in one graph or
highlight each region in separate graphs. Here we will use the
terminology “spatial domain” and “regions” interchangeably.</p>
<p>Depending on the number of regions, it may be more useful to
visualise the spatial regions either collectively in a single graph or
separately in multiple graphs. To visualise it in a single graph, the
<code><a href="https://rdrr.io/pkg/lisaClust/man/hatchingPlot.html" class="external-link">hatchingPlot()</a></code> function is used to create hatching patterns
to represent spatial regions and cell-types. The hatching geom is used
to create hatching patterns to represent spatial regions.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## To visualise it in a single graph</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/hatchingPlot.html" class="external-link">hatchingPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">data_sce</span>,</span>
<span>  useImages <span class="op">=</span> <span class="st">"IZ_P9_cond_Ischaemic"</span>, </span>
<span>  region <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>  imageID <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"celltype"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"spatial_x"</span>, <span class="st">"spatial_y"</span><span class="op">)</span> <span class="op">)</span> </span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-12-1.png" width="576"></p>
<p>We have written a small function
<code>draw_region_clustering_result</code> to visualise the data
separately in multiple graphs for the individual
<code>IZ_P9_cond_Ischaemic</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw_region_clustering_result</span><span class="op">(</span><span class="va">data_sce</span> , </span>
<span>                              sample <span class="op">=</span> <span class="st">"sample"</span> , </span>
<span>                              selected_sample <span class="op">=</span>  <span class="st">"IZ_P9_cond_Ischaemic"</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span></code></pre>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-13-1.png" width="960"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span></code></pre>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-13-2.png" width="960"></p>
</div>
<div class="section level4">
<h4 class="unnumbered" id="across-individuals">Across individuals<a class="anchor" aria-label="anchor" href="#across-individuals"></a>
</h4>
<p>We can better interpret the region outputs by summarising the
proportion of each cell type in a region across the individuals.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw_dotplot</span><span class="op">(</span><span class="va">data_sce</span>,  </span>
<span>             sample <span class="op">=</span> <span class="st">"sample"</span> , </span>
<span>             celltype <span class="op">=</span> <span class="st">"celltype"</span> ,  </span>
<span>             group<span class="op">=</span> <span class="st">"condition"</span> , </span>
<span>             group_of_interest <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myogenic"</span> , <span class="st">"Ischaemic"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-14-1.png" width="960"></p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i>
<i class="far fa-chart-bar fa-stack-1x fa-inverse"></i> </span>
<strong>Critical Thinking Q:</strong></p>
<p>Q4: Which regions would you focus on next?</p>
</div>
<p>[optional] A complex graphical options using ggplot</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span> <span class="va">data_sce</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">sample</span> , <span class="va">condition</span> , <span class="va">region</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">celltype</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>proportion <span class="op">=</span> <span class="va">n</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">proportion</span>, x <span class="op">=</span> <span class="va">sample</span> , fill <span class="op">=</span> <span class="va">celltype</span><span class="op">)</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="op">~</span><span class="va">region</span><span class="op">+</span><span class="va">condition</span>, scales <span class="op">=</span> <span class="st">"free"</span>, space <span class="op">=</span> <span class="st">"free"</span> <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">color_codes</span><span class="op">)</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.title.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.text.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.ticks.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-15-1.png" width="960"></p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i>
<i class="far fa-chart-bar fa-stack-1x fa-inverse"></i> </span>
<strong>Critical Thinking Q:</strong></p>
<p>Q5: Does your conclusion change after looking at a different
plot?</p>
</div>
</div>
<div class="section level4">
<h4 class="unnumbered" id="further-exploration-by-visualising-selected-regions">Further exploration by visualising selected
regions<a class="anchor" aria-label="anchor" href="#further-exploration-by-visualising-selected-regions"></a>
</h4>
<p>The number of sub-cell types increase considerably when we want to
add spatial domain (region) information. To enhance clarity and
facilitate understanding, it may be helpful to choose a predetermined
region. The code generates a set of boxplots that enable the comparison
of cell type proportions between individuals with ischaemic and myogenic
conditions in <code>region_2</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw_selected_region_boxplot</span><span class="op">(</span><span class="va">data_sce</span>, </span>
<span>                              sample <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span>                              celltype <span class="op">=</span> <span class="st">"celltype"</span> ,  </span>
<span>                              group<span class="op">=</span> <span class="st">"condition"</span> , </span>
<span>                              group_of_interest <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myogenic"</span> , <span class="st">"Ischaemic"</span><span class="op">)</span> , </span>
<span>                             select_region <span class="op">=</span> <span class="st">"region_2"</span> <span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-16-1.png" width="1152"></p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<p>Comparing cell-type compositions in the region, what can you tell
about the regions?</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="how-do-we-generate-molecular-representations-for-each-individual">How do we generate molecular representations for each
individual?<a class="anchor" aria-label="anchor" href="#how-do-we-generate-molecular-representations-for-each-individual"></a>
</h2>
<p>In this workshop, we use <code>scFeatures</code> to generate
molecular representations for each patient. The molecular representation
is interpretable and hence facilitates downstream analysis of the
patient. Overall, <code>scFeatures</code> generates features across six
categories representing different molecular views of cellular
characteristics. These include: - i) cell type proportions - ii) cell
type specific gene expressions - iii) cell type specific pathway
expressions - iv) cell type specific cell-cell interaction (CCI) scores
- v) overall aggregated gene expressions - vi) spatial metrics The
different types of features constructed enable a more comprehensive
multi-view understanding of each patient from a matrix of
<code>spots x cells</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"number of cells in each sample"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "number of cells in each sample"</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  control_P1_cond_Myogenic control_P17_cond_Myogenic  control_P7_cond_Myogenic </span></span>
<span><span class="co">##                      1000                      1000                      1000 </span></span>
<span><span class="co">##  control_P8_cond_Myogenic     IZ_P10_cond_Ischaemic     IZ_P13_cond_Ischaemic </span></span>
<span><span class="co">##                      1000                      1000                      1000 </span></span>
<span><span class="co">##     IZ_P15_cond_Ischaemic     IZ_P16_cond_Ischaemic      IZ_P2_cond_Ischaemic </span></span>
<span><span class="co">##                      1000                      1000                      1000 </span></span>
<span><span class="co">##      IZ_P3_cond_Ischaemic      IZ_P9_cond_Ischaemic IZ_P9_rep2_cond_Ischaemic </span></span>
<span><span class="co">##                      1000                      1000                      1000 </span></span>
<span><span class="co">##      RZ_P11_cond_Myogenic      RZ_P12_cond_Myogenic       RZ_P2_cond_Myogenic </span></span>
<span><span class="co">##                      1000                      1000                      1000 </span></span>
<span><span class="co">##       RZ_P3_cond_Myogenic       RZ_P5_cond_Myogenic       RZ_P6_cond_Myogenic </span></span>
<span><span class="co">##                      1000                      1000                      1000 </span></span>
<span><span class="co">##       RZ_P9_cond_Myogenic </span></span>
<span><span class="co">##                      1000</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># metadata &lt;- data.frame( table(data_sce$sample) )</span></span>
<span><span class="co"># colnames(metadata)[1] &lt;- "Patient"</span></span>
<span><span class="co"># DT::datatable(metadata , options = list(pageLength = 5), width = "400px")</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"number of cells in each celltype"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "number of cells in each celltype"</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Cardiomyocyte Cycling.cells   Endothelial    Fibroblast          Mast </span></span>
<span><span class="co">##          8782          1755          1878          3714            41 </span></span>
<span><span class="co">##       Myeloid      Pericyte         vSMCs      Lymphoid      Neuronal </span></span>
<span><span class="co">##          2421            44           266            40            34 </span></span>
<span><span class="co">##     Adipocyte </span></span>
<span><span class="co">##            25</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># metadata &lt;- data.frame( table(data_sce$celltype) )</span></span>
<span><span class="co"># colnames(metadata)[1] &lt;- "Region specific cell type"</span></span>
<span><span class="co"># DT::datatable(metadata , options = list(pageLength = 5), width = "400px")</span></span>
<span><span class="co">#   </span></span></code></pre></div>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<p>Are there any samples or cell types you would like to remove from the
data?</p>
</div>
<div class="section level3">
<h3 class="tabset" id="how-to-create-molecular-representations-of-individuals">How to create molecular representations of
individuals ?<a class="anchor" aria-label="anchor" href="#how-to-create-molecular-representations-of-individuals"></a>
</h3>
<p>There are different ways you can use <code>scFeatures</code> to
generate molecular representations for individuals and it requires the
following information for spatial data.</p>
<ul>
<li>data,<br>
</li>
<li>sample,</li>
<li>X coordinates,</li>
<li>Y coordinates,</li>
<li>feature_types, and</li>
<li>type</li>
</ul>
<p>There are a total of 13 different types of features (feature_types)
that you can choose to generate. The arugment type refers the type of
input data we have. This is either <code>scrna</code> (single-cell
RNA-sequencing data), <code>spatial_p</code> (spatial proteomics data),
or <code>spatial_t</code> (single-cell spatial data).</p>
<div class="section level4">
<h4 class="unnumbered" id="calculate-feature_types">Calculate feature_types<a class="anchor" aria-label="anchor" href="#calculate-feature_types"></a>
</h4>
<p>Suppose that we are interested in determining the proportion of each
cell type in each individual within each region. It is necessary to
specify <code>type = spatial_t</code> to reflect that we have Visum data
and <code>feature_types = proportion_raw</code> to indicate we intend to
calculate cell type proportion for each of the region-specific sub-cell
types.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## [A] The next few lines extract specific information from data_sce as input to scFeatures. </span></span>
<span></span>
<span><span class="co">## Extract the expression matrix from data_sce</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span> <span class="va">data_sce</span> <span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract the sample information </span></span>
<span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">data_sce</span><span class="op">$</span><span class="va">sample</span></span>
<span></span>
<span><span class="co">## Extract the x, y coordiantes</span></span>
<span><span class="va">spatialCoords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">$</span><span class="va">spatial_x</span>, <span class="va">data_sce</span><span class="op">$</span><span class="va">spatial_y</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># For spatial transcriptomics, we need to specify the cell type prediction of each spot</span></span>
<span><span class="va">prediction.scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">data_sce</span><span class="op">)</span><span class="op">[</span>, <span class="va">celltype</span><span class="op">]</span></span>
<span><span class="va">prediction.scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">prediction.scores</span><span class="op">)</span>  <span class="op">)</span></span>
<span></span>
<span><span class="co">### [B] Running scFeatures</span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/scFeatures/reference/scFeatures.html" class="external-link">scFeatures</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">data</span> ,  </span>
<span>                                sample <span class="op">=</span> <span class="va">sample</span>, </span>
<span>                                spatialCoords <span class="op">=</span> <span class="va">spatialCoords</span>, </span>
<span>                                spotProbability <span class="op">=</span> <span class="va">prediction.scores</span>, </span>
<span>                                feature_types <span class="op">=</span> <span class="st">"proportion_raw"</span>,</span>
<span>                                type <span class="op">=</span><span class="st">"spatial_t"</span> <span class="op">)</span></span></code></pre></div>
<p>The generated feature is a matrix of <code>samples x features</code>,
the code below shows a compositional plot for each individuals</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">proportion_raw</span></span>
<span><span class="fu">plot_barplot</span><span class="op">(</span> <span class="va">feature</span> <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Proportion raw feature"</span><span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-19-1.png" width="1920"></p>
</div>
<div class="section level4">
<h4 class="unnumbered" id="all-cell-types-and-features-in-one-line-of-code">All cell types and features in one line of
code<a class="anchor" aria-label="anchor" href="#all-cell-types-and-features-in-one-line-of-code"></a>
</h4>
<p>The code below enable us to generate all feature types for all cell
types in a line. Due to limitations with today’s computational capacity,
** Please DO NOT run it in today’s workshop, it will crash your
system**.</p>
<pre><code><span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/scFeatures/reference/scFeatures.html" class="external-link">scFeatures</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">data</span> ,  </span>
<span>                                sample <span class="op">=</span> <span class="va">sample</span>, </span>
<span>                                spatialCoords <span class="op">=</span> <span class="va">spatialCoords</span> , </span>
<span>                                spotProbability <span class="op">=</span> <span class="va">prediction.scores</span>, </span>
<span>                                type <span class="op">=</span><span class="st">"spatial_t"</span>, </span>
<span>                                ncores <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/biocasia_data/scfeatures_result_matrix.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="what-can-we-learn-from-the-outputs-from-scfeatures">What can we learn from the outputs from scFeatures?<a class="anchor" aria-label="anchor" href="#what-can-we-learn-from-the-outputs-from-scfeatures"></a>
</h3>
<p>Assuming you have already generated a collection of molecular
representations for individuals. Please load the prepared RDS file
<code>scfeatures_result_matrix.rds</code>. Again, you can remind
yourself that all generated feature types are stored in a matrix of
<code>samples x features</code>.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Upload pre-generated RDS file</span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/biocasia_data/scfeatures_result_matrix.rds"</span><span class="op">)</span></span>
<span> </span>
<span><span class="co"># We have generated a total of 13 types of feature categories</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "proportion_raw"       "proportion_logit"     "proportion_ratio"    </span></span>
<span><span class="co">##  [4] "gene_mean_celltype"   "pathway_mean"         "gene_mean_bulk"      </span></span>
<span><span class="co">##  [7] "gene_prop_bulk"       "gene_cor_bulk"        "L_stats"             </span></span>
<span><span class="co">## [10] "celltype_interaction" "morans_I"             "nn_correlation"</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># What is the number of features that we have generated </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">scfeatures_result</span>, <span class="va">dim</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $proportion_raw</span></span>
<span><span class="co">## [1] 19 11</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $proportion_logit</span></span>
<span><span class="co">## [1] 19 11</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $proportion_ratio</span></span>
<span><span class="co">## [1] 19 55</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_mean_celltype</span></span>
<span><span class="co">## [1]    19 33561</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pathway_mean</span></span>
<span><span class="co">## [1]  19 550</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_mean_bulk</span></span>
<span><span class="co">## [1]   19 3051</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_prop_bulk</span></span>
<span><span class="co">## [1]   19 3051</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gene_cor_bulk</span></span>
<span><span class="co">## [1]    19 60031</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $L_stats</span></span>
<span><span class="co">## [1]  19 121</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $celltype_interaction</span></span>
<span><span class="co">## [1]  19 121</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $morans_I</span></span>
<span><span class="co">## [1]   19 3051</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $nn_correlation</span></span>
<span><span class="co">## [1]   19 3051</span></span></code></pre>
<div class="section level4">
<h4 id="association-study---can-we-identify-differential-expression-for-a-feature-of-interest">Association study - can we identify “differential expression” for a
feature of interest<a class="anchor" aria-label="anchor" href="#association-study---can-we-identify-differential-expression-for-a-feature-of-interest"></a>
</h4>
<p>The R object <code>scfeatures_result</code> contains a variety of
features. An important question focuses on the identification of
features that reflect an association with the prognostic outcome,
specifically distinguishing between disease conditions. The code
provided below demonstrates the use of the <code>limma()</code> function
to fit a linear model for the purpose of analysing gene_mean_celltype as
an illustration feature. The feature type known as
<code>gene_mean_celltype</code> represents the mean gene expression for
each sub-cell type specific to a spatial region.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract cell type specific gene expression for all regions. </span></span>
<span><span class="va">gene_mean_bulk</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">gene_mean_bulk</span></span>
<span></span>
<span><span class="co"># this transposes the data </span></span>
<span><span class="co"># in bioinformatics convention, features are stored in rows </span></span>
<span><span class="co"># in statistics convention, features are stored in columns</span></span>
<span><span class="va">gene_mean_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">gene_mean_bulk</span><span class="op">)</span></span>
<span>      </span>
<span></span>
<span><span class="co"># Extract the two conditions of interest - disease condition</span></span>
<span><span class="va">condition</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gene_mean_bulk</span><span class="op">)</span>, <span class="st">"_cond_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>condition <span class="op">=</span> <span class="va">condition</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate log fold change each gene using limma</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">condition</span>, data <span class="op">=</span> <span class="va">condition</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">gene_mean_bulk</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span><span class="va">tT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">fit</span>, n <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span>
<span><span class="va">tT</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tT</span><span class="op">)</span></span></code></pre></div>
<p>We visualise the comparison using a volcano plot and a dotplot for
the cell type specific expression feature. This is a type of
scatter-plot that is used to quickly identify changes in large data sets
and represent the significance (y-axis) versus effect size or
fold-change (x-axis).</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span> <span class="va">tT</span> , <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">logFC</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">P.Value</span><span class="op">)</span> , text <span class="op">=</span> <span class="va">gene</span> <span class="op">)</span> <span class="op">)</span><span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">P.Value</span><span class="op">)</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, size<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"blue"</span>,high<span class="op">=</span><span class="st">"red"</span><span class="op">)</span><span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"log2 fold change"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"-log10 p-value"</span><span class="op">)</span>  </span>
<span></span>
<span><span class="va">tT</span> <span class="op">&lt;-</span> <span class="va">tT</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">tT</span><span class="op">$</span><span class="va">logFC</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">tT</span> <span class="op">&lt;-</span> <span class="va">tT</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="op">]</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span> <span class="va">tT</span> , <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span> y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">gene</span>, <span class="va">logFC</span><span class="op">)</span> , x <span class="op">=</span> <span class="va">logFC</span>  <span class="op">)</span> <span class="op">)</span><span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">P.Value</span><span class="op">)</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, size<span class="op">=</span><span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"blue"</span>,high<span class="op">=</span><span class="st">"red"</span><span class="op">)</span><span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"logFC"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"region specific cel type specfic features"</span> <span class="op">)</span> </span>
<span> </span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>,<span class="va">p2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-23-1.png" width="672"></p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Critical Thinking Q:</strong></p>
<p>Q6: Which figure do you prefer? The volcano plot or the dotplot?</p>
</div>
</div>
<div class="section level4">
<h4 class="tabset" id="can-we-identify-feature-of-interest-among-each-types-of-feature-category">Can we identify feature of interest among each types
of feature category?<a class="anchor" aria-label="anchor" href="#can-we-identify-feature-of-interest-among-each-types-of-feature-category"></a>
</h4>
<p>In order to further develop our understanding of our data, it is
important to bear in mind that the outcomes derived from
<code>scFeatures</code> combine several association studies. Hence, it
is necessary to examine every separate feature category as an individual
unit of study. This section presents a graphic representation showing
how to visualise the distribution of features across individuals. The
following code has been developed to identify and retrieve all samples
that have been associated to disease outcome. This particular step is
necessary in this setting as our focus is limited to two specific
categories of individuals rather than including all samples.</p>
<div class="section level5">
<h5 class="unnumbered" id="proportion-raw">Proportion raw<a class="anchor" aria-label="anchor" href="#proportion-raw"></a>
</h5>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">proportion_raw</span></span>
<span> </span>
<span><span class="fu">plot_barplot</span><span class="op">(</span><span class="va">feature</span> <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Proportion raw feature"</span><span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="gene-mean-celltype">Gene mean celltype<a class="anchor" aria-label="anchor" href="#gene-mean-celltype"></a>
</h5>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">gene_mean_celltype</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span>
<span><span class="va">feature</span>   <span class="op">&lt;-</span> <span class="va">feature</span><span class="op">[</span>  , <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Cardio"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span><span class="op">)</span> <span class="op">]</span></span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Gene mean celltype feature"</span><span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="pathway-mean">Pathway mean<a class="anchor" aria-label="anchor" href="#pathway-mean"></a>
</h5>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">pathway_mean</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Pathway mean feature"</span><span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="gene-mean-bulk">Gene mean bulk<a class="anchor" aria-label="anchor" href="#gene-mean-bulk"></a>
</h5>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">gene_mean_bulk</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span></span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Gene mean bulk feature"</span><span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="nearest-neighbour-correlation">Nearest neighbour correlation<a class="anchor" aria-label="anchor" href="#nearest-neighbour-correlation"></a>
</h5>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">nn_correlation</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="fu">plot_boxplot</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Nearest neighbour correlation feature"</span><span class="op">)</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
</div>
</div>
<div class="section level4">
<h4 id="association-study-report---are-the-generated-features-sensible">Association study report - are the generated features sensible?<a class="anchor" aria-label="anchor" href="#association-study-report---are-the-generated-features-sensible"></a>
</h4>
<p>To accommodate easier interpretation of the features,
<code>scFeatures</code> contains a function
<code>run_association_study_report</code> that enables the user to
readily visualise and explore all generated features with one line of
code. This should be used with caution as depending on the number of
cell-types considered, some of the automatically-generated graphics may
not be the most efficient way for interpretation. Specify a folder to
store the html report. Using the function <code><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd()</a></code>, we
specify that we will store it in the current working directory.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_folder</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/scFeatures/reference/run_association_study_report.html" class="external-link">run_association_study_report</a></span><span class="op">(</span><span class="va">scfeatures_result</span>, <span class="va">output_folder</span> <span class="op">)</span></span></code></pre></div>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Critical Thinking Q:</strong></p>
<p>Using the HTML, we can look at some of the critical thinking
questions that a researcher would ask about the generated features.
These questions are exploratory and there is no right or wrong
answer.</p>
<p>Q7: Do the generated features look reasonable?<br>
Which cell type(s) would you like to focus on at your next stage of
analysis?<br>
Which feature type(s) would you like to focus on at your next stage of
analysis?<br>
Q8: Are the conditions in your data relatively easy or difficult to
distinguish?</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="can-we-classify-or-discrimiante-between-the-ischaemic-and-myogenic-samples">Can we classify or discrimiante between the ischaemic and myogenic
samples?<a class="anchor" aria-label="anchor" href="#can-we-classify-or-discrimiante-between-the-ischaemic-and-myogenic-samples"></a>
</h2>
<p>In this section we build a classification model to predict whether
the samples belong to the ischaemic or myogenic group.</p>
<div class="section level3">
<h3 id="building-classification-model">Building classification model<a class="anchor" aria-label="anchor" href="#building-classification-model"></a>
</h3>
<p>Recall in the previous section that we have stored the 13 feature
types matrix in a list. Instead of manually retrieving each matrix from
the list to build separate models, <code>classifyR</code> can directly
take a list of matrices as an input and run a repeated cross-validation
model on each matrix individually.</p>
<p>Below, we run 5 repeats of 3 folds cross-validation.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outcome</span> <span class="op">=</span> <span class="va">scfeatures_result</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">rownames</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="st">"_cond_"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### generate classfyr result </span></span>
<span><span class="va">classifyr_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span><span class="va">scfeatures_result</span>,</span>
<span>                                 <span class="va">outcome</span>, </span>
<span>                                 classifier <span class="op">=</span> <span class="st">"kNN"</span>,</span>
<span>                                 nFolds <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                                 nRepeats <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                                 nCores <span class="op">=</span> <span class="fl">5</span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualising-the-classification-performance">Visualising the classification performance<a class="anchor" aria-label="anchor" href="#visualising-the-classification-performance"></a>
</h3>
<p>To examine the classification model performance, we first need to
specify a metric to calculate. Here, we calculate the balanced
accuracy.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">classifyr_result</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/biocasia_data/classifyr_result.rds"</span><span class="op">)</span></span>
<span><span class="va">classifyr_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">classifyr_result</span>, </span>
<span>                           <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/calcPerformance.html" class="external-link">calcCVperformance</a></span><span class="op">(</span><span class="va">x</span>, performanceType <span class="op">=</span> <span class="st">"Balanced Accuracy"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Format the output and visualise the accuracy using boxplots.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">level_order</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html" class="external-link">performancePlot</a></span><span class="op">(</span><span class="va">classifyr_result</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_x_discrete</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="va">level_order</span><span class="op">)</span>  </span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="VisiumVersion3_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h3>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Debian GNU/Linux 12 (bookworm)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.21.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Australia/Sydney</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] reshape_0.8.9               SPOTlight_1.5.2            </span></span>
<span><span class="co">##  [3] scran_1.28.1                scater_1.28.0              </span></span>
<span><span class="co">##  [5] scuttle_1.10.1              spatstat.geom_3.2-7        </span></span>
<span><span class="co">##  [7] spatstat.data_3.0-3         survminer_0.4.9            </span></span>
<span><span class="co">##  [9] ggpubr_0.6.0                tidyr_1.3.0                </span></span>
<span><span class="co">## [11] scattermore_1.2             plotly_4.10.3              </span></span>
<span><span class="co">## [13] limma_3.56.2                dplyr_1.1.3                </span></span>
<span><span class="co">## [15] spicyR_1.12.0               ggthemes_4.2.4             </span></span>
<span><span class="co">## [17] lisaClust_1.9.2             ClassifyR_3.5.21           </span></span>
<span><span class="co">## [19] survival_3.5-7              BiocParallel_1.34.2        </span></span>
<span><span class="co">## [21] MultiAssayExperiment_1.26.0 generics_0.1.3             </span></span>
<span><span class="co">## [23] scFeatures_0.99.27          ggplot2_3.4.4              </span></span>
<span><span class="co">## [25] SpatialExperiment_1.11.2    SingleCellExperiment_1.22.0</span></span>
<span><span class="co">## [27] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">## [29] GenomicRanges_1.52.0        GenomeInfoDb_1.36.1        </span></span>
<span><span class="co">## [31] IRanges_2.34.1              S4Vectors_0.38.1           </span></span>
<span><span class="co">## [33] BiocGenerics_0.46.0         MatrixGenerics_1.12.2      </span></span>
<span><span class="co">## [35] matrixStats_1.0.0          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] R.methodsS3_1.8.2          GSEABase_1.62.0           </span></span>
<span><span class="co">##   [3] progress_1.2.2             tiff_0.1-11               </span></span>
<span><span class="co">##   [5] EnsDb.Mmusculus.v79_2.99.0 goftest_1.2-3             </span></span>
<span><span class="co">##   [7] DT_0.30                    Biostrings_2.68.1         </span></span>
<span><span class="co">##   [9] HDF5Array_1.28.1           vctrs_0.6.4               </span></span>
<span><span class="co">##  [11] spatstat.random_3.2-1      digest_0.6.33             </span></span>
<span><span class="co">##  [13] png_0.1-8                  shape_1.4.6               </span></span>
<span><span class="co">##  [15] EnsDb.Hsapiens.v79_2.99.0  registry_0.5-1            </span></span>
<span><span class="co">##  [17] ggrepel_0.9.4              deldir_1.0-9              </span></span>
<span><span class="co">##  [19] magick_2.8.1               MASS_7.3-60               </span></span>
<span><span class="co">##  [21] pkgdown_2.0.7              reshape2_1.4.4            </span></span>
<span><span class="co">##  [23] httpuv_1.6.12              foreach_1.5.2             </span></span>
<span><span class="co">##  [25] withr_2.5.1                ggfun_0.1.3               </span></span>
<span><span class="co">##  [27] xfun_0.40                  ellipsis_0.3.2            </span></span>
<span><span class="co">##  [29] cytomapper_1.12.0          memoise_2.0.1             </span></span>
<span><span class="co">##  [31] proxyC_0.3.4               ggbeeswarm_0.7.2          </span></span>
<span><span class="co">##  [33] systemfonts_1.0.5          ragg_1.2.6                </span></span>
<span><span class="co">##  [35] zoo_1.8-12                 GlobalOptions_0.1.2       </span></span>
<span><span class="co">##  [37] gtools_3.9.4               V8_4.4.0                  </span></span>
<span><span class="co">##  [39] SingleCellSignalR_1.12.0   R.oo_1.25.0               </span></span>
<span><span class="co">##  [41] prettyunits_1.2.0          KEGGREST_1.40.0           </span></span>
<span><span class="co">##  [43] promises_1.2.1             httr_1.4.7                </span></span>
<span><span class="co">##  [45] rstatix_0.7.2              restfulr_0.0.15           </span></span>
<span><span class="co">##  [47] rhdf5filters_1.12.1        rhdf5_2.44.0              </span></span>
<span><span class="co">##  [49] rstudioapi_0.15.0          concaveman_1.1.0          </span></span>
<span><span class="co">##  [51] babelgene_22.9             curl_5.0.2                </span></span>
<span><span class="co">##  [53] zlibbioc_1.46.0            ScaledMatrix_1.8.1        </span></span>
<span><span class="co">##  [55] polyclip_1.10-6            GenomeInfoDbData_1.2.10   </span></span>
<span><span class="co">##  [57] fftwtools_0.9-11           xtable_1.8-4              </span></span>
<span><span class="co">##  [59] stringr_1.5.0              desc_1.4.2                </span></span>
<span><span class="co">##  [61] doParallel_1.0.17          evaluate_0.22             </span></span>
<span><span class="co">##  [63] S4Arrays_1.0.6             BiocFileCache_2.8.0       </span></span>
<span><span class="co">##  [65] hms_1.1.3                  irlba_2.3.5.1             </span></span>
<span><span class="co">##  [67] colorspace_2.1-0           filelock_1.0.2            </span></span>
<span><span class="co">##  [69] magrittr_2.0.3             later_1.3.1               </span></span>
<span><span class="co">##  [71] viridis_0.6.4              lattice_0.22-5            </span></span>
<span><span class="co">##  [73] NMF_0.26                   cowplot_1.1.1             </span></span>
<span><span class="co">##  [75] XML_3.99-0.14              ggupset_0.3.0             </span></span>
<span><span class="co">##  [77] RcppAnnoy_0.0.21           svgPanZoom_0.3.4          </span></span>
<span><span class="co">##  [79] class_7.3-22               pillar_1.9.0              </span></span>
<span><span class="co">##  [81] simpleSeg_1.2.0            nlme_3.1-163              </span></span>
<span><span class="co">##  [83] EBImage_4.42.0             iterators_1.0.14          </span></span>
<span><span class="co">##  [85] gridBase_0.4-7             caTools_1.18.2            </span></span>
<span><span class="co">##  [87] compiler_4.3.1             beachmat_2.16.0           </span></span>
<span><span class="co">##  [89] stringi_1.7.12             tensor_1.5                </span></span>
<span><span class="co">##  [91] minqa_1.2.6                GenomicAlignments_1.36.0  </span></span>
<span><span class="co">##  [93] plyr_1.8.9                 msigdbr_7.5.1             </span></span>
<span><span class="co">##  [95] crayon_1.5.2               abind_1.4-5               </span></span>
<span><span class="co">##  [97] BiocIO_1.10.0              sp_2.1-1                  </span></span>
<span><span class="co">##  [99] locfit_1.5-9.8             terra_1.7-55              </span></span>
<span><span class="co">## [101] bit_4.0.5                  codetools_0.2-19          </span></span>
<span><span class="co">## [103] textshaping_0.3.7          BiocSingular_1.16.0       </span></span>
<span><span class="co">## [105] crosstalk_1.2.0            bslib_0.5.1               </span></span>
<span><span class="co">## [107] multtest_2.56.0            mime_0.12                 </span></span>
<span><span class="co">## [109] splines_4.3.1              circlize_0.4.15           </span></span>
<span><span class="co">## [111] Rcpp_1.0.11                dbplyr_2.4.0              </span></span>
<span><span class="co">## [113] sparseMatrixStats_1.12.2   knitr_1.44                </span></span>
<span><span class="co">## [115] blob_1.2.4                 utf8_1.2.4                </span></span>
<span><span class="co">## [117] AnnotationFilter_1.24.0    lme4_1.1-34               </span></span>
<span><span class="co">## [119] fs_1.6.3                   nnls_1.5                  </span></span>
<span><span class="co">## [121] DelayedMatrixStats_1.22.1  GSVA_1.48.2               </span></span>
<span><span class="co">## [123] ggsignif_0.6.4             tibble_3.2.1              </span></span>
<span><span class="co">## [125] Matrix_1.6-1.1             scam_1.2-14               </span></span>
<span><span class="co">## [127] statmod_1.5.0              svglite_2.1.2             </span></span>
<span><span class="co">## [129] tweenr_2.0.2               pkgconfig_2.0.3           </span></span>
<span><span class="co">## [131] pheatmap_1.0.12            tools_4.3.1               </span></span>
<span><span class="co">## [133] cachem_1.0.8               RSQLite_2.3.1             </span></span>
<span><span class="co">## [135] viridisLite_0.4.2          DBI_1.1.3                 </span></span>
<span><span class="co">## [137] numDeriv_2016.8-1.1        fastmap_1.1.1             </span></span>
<span><span class="co">## [139] rmarkdown_2.25             scales_1.2.1              </span></span>
<span><span class="co">## [141] grid_4.3.1                 shinydashboard_0.7.2      </span></span>
<span><span class="co">## [143] Rsamtools_2.16.0           broom_1.0.5               </span></span>
<span><span class="co">## [145] sass_0.4.7                 BiocManager_1.30.22       </span></span>
<span><span class="co">## [147] graph_1.78.0               carData_3.0-5             </span></span>
<span><span class="co">## [149] farver_2.1.1               scatterpie_0.2.1          </span></span>
<span><span class="co">## [151] mgcv_1.9-0                 yaml_2.3.7                </span></span>
<span><span class="co">## [153] rtracklayer_1.60.0         cli_3.6.1                 </span></span>
<span><span class="co">## [155] purrr_1.0.2                lifecycle_1.0.3           </span></span>
<span><span class="co">## [157] uwot_0.1.16                bluster_1.10.0            </span></span>
<span><span class="co">## [159] backports_1.4.1            annotate_1.78.0           </span></span>
<span><span class="co">## [161] gtable_0.3.4               rjson_0.2.21              </span></span>
<span><span class="co">## [163] parallel_4.3.1             ape_5.7-1                 </span></span>
<span><span class="co">## [165] jsonlite_1.8.7             edgeR_3.42.4              </span></span>
<span><span class="co">## [167] bitops_1.0-7               bit64_4.0.5               </span></span>
<span><span class="co">## [169] Rtsne_0.16                 spatstat.utils_3.0-4      </span></span>
<span><span class="co">## [171] BiocNeighbors_1.18.0       RcppParallel_5.1.7        </span></span>
<span><span class="co">## [173] jquerylib_0.1.4            metapod_1.8.0             </span></span>
<span><span class="co">## [175] dqrng_0.3.1                survMisc_0.5.6            </span></span>
<span><span class="co">## [177] R.utils_2.12.2             lazyeval_0.2.2            </span></span>
<span><span class="co">## [179] shiny_1.7.5.1              htmltools_0.5.6.1         </span></span>
<span><span class="co">## [181] KMsurv_0.1-5               rappdirs_0.3.3            </span></span>
<span><span class="co">## [183] ensembldb_2.24.0           glue_1.6.2                </span></span>
<span><span class="co">## [185] XVector_0.40.0             RCurl_1.98-1.12           </span></span>
<span><span class="co">## [187] rprojroot_2.0.3            jpeg_0.1-10               </span></span>
<span><span class="co">## [189] gridExtra_2.3              AUCell_1.22.0             </span></span>
<span><span class="co">## [191] boot_1.3-28.1              igraph_1.5.1              </span></span>
<span><span class="co">## [193] R6_2.5.1                   gplots_3.1.3              </span></span>
<span><span class="co">## [195] labeling_0.4.3             km.ci_0.5-6               </span></span>
<span><span class="co">## [197] GenomicFeatures_1.52.1     cluster_2.1.4             </span></span>
<span><span class="co">## [199] rngtools_1.5.2             Rhdf5lib_1.22.0           </span></span>
<span><span class="co">## [201] nloptr_2.0.3               DelayedArray_0.26.6       </span></span>
<span><span class="co">## [203] tidyselect_1.2.0           vipor_0.4.5               </span></span>
<span><span class="co">## [205] ProtGenerics_1.32.0        raster_3.6-26             </span></span>
<span><span class="co">## [207] ggforce_0.4.1              xml2_1.3.5                </span></span>
<span><span class="co">## [209] car_3.1-2                  AnnotationDbi_1.62.2      </span></span>
<span><span class="co">## [211] rsvd_1.0.5                 munsell_0.5.0             </span></span>
<span><span class="co">## [213] KernSmooth_2.23-22         data.table_1.14.8         </span></span>
<span><span class="co">## [215] htmlwidgets_1.6.2          RColorBrewer_1.1-3        </span></span>
<span><span class="co">## [217] biomaRt_2.56.1             rlang_1.1.1               </span></span>
<span><span class="co">## [219] spatstat.sparse_3.0-3      spatstat.explore_3.2-5    </span></span>
<span><span class="co">## [221] lmerTest_3.1-3             fansi_1.0.5               </span></span>
<span><span class="co">## [223] beeswarm_0.4.0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="acknowledgment">Acknowledgment<a class="anchor" aria-label="anchor" href="#acknowledgment"></a>
</h3>
<p>The authors thank all their colleagues, particularly at The
University of Sydney, Sydney Precision Data Science and Charles Perkins
Centre for their support and intellectual engagement. Special thanks to
Ellis Patrick, Shila Ghazanfar, Andy Tran and Helen Fu for guiding and
supporting the building of this workshop.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Sydney Precision Bioinformatics Alliance.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
